{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Data wrangling and curation of Fermentos dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd # type: ignore"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>mag_id</th>\n",
       "      <th>sample_description</th>\n",
       "      <th>fermented_food</th>\n",
       "      <th>specific_substrate</th>\n",
       "      <th>substrate_category</th>\n",
       "      <th>general_category</th>\n",
       "      <th>completeness</th>\n",
       "      <th>contamination</th>\n",
       "      <th>contigs</th>\n",
       "      <th>total_length</th>\n",
       "      <th>...</th>\n",
       "      <th>n50</th>\n",
       "      <th>domain</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>order</th>\n",
       "      <th>family</th>\n",
       "      <th>genus</th>\n",
       "      <th>species</th>\n",
       "      <th>study_catalog</th>\n",
       "      <th>source</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AlvarezOrdonezA_xxxx__C02-18__bin.2</td>\n",
       "      <td>meat</td>\n",
       "      <td>sausage</td>\n",
       "      <td>meat</td>\n",
       "      <td>meat</td>\n",
       "      <td>fermented_meat</td>\n",
       "      <td>51.600000</td>\n",
       "      <td>3.100000</td>\n",
       "      <td>2506</td>\n",
       "      <td>20998336</td>\n",
       "      <td>...</td>\n",
       "      <td>9910</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Ascomycota</td>\n",
       "      <td>Eurotiomycetes</td>\n",
       "      <td>Eurotiales</td>\n",
       "      <td>Aspergillaceae</td>\n",
       "      <td>Aspergillus</td>\n",
       "      <td>Aspergillus_glaucus</td>\n",
       "      <td>Carlino2024</td>\n",
       "      <td>AlvarezOrdonezA_xxxx__C02-18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AlvarezOrdonezA_xxxx__C02-19__bin.5</td>\n",
       "      <td>meat</td>\n",
       "      <td>sausage</td>\n",
       "      <td>meat</td>\n",
       "      <td>meat</td>\n",
       "      <td>fermented_meat</td>\n",
       "      <td>63.793103</td>\n",
       "      <td>0.862069</td>\n",
       "      <td>542</td>\n",
       "      <td>1596974</td>\n",
       "      <td>...</td>\n",
       "      <td>3152</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Firmicutes</td>\n",
       "      <td>Bacilli</td>\n",
       "      <td>Bacillales</td>\n",
       "      <td>Staphylococcaceae</td>\n",
       "      <td>Staphylococcus</td>\n",
       "      <td>Staphylococcus_equorum</td>\n",
       "      <td>Carlino2024</td>\n",
       "      <td>AlvarezOrdonezA_xxxx__C02-19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AlvarezOrdonezA_xxxx__C02-19__bin.3</td>\n",
       "      <td>meat</td>\n",
       "      <td>sausage</td>\n",
       "      <td>meat</td>\n",
       "      <td>meat</td>\n",
       "      <td>fermented_meat</td>\n",
       "      <td>74.137931</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>279</td>\n",
       "      <td>5340756</td>\n",
       "      <td>...</td>\n",
       "      <td>30953</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Pseudomonadales</td>\n",
       "      <td>Pseudomonadaceae</td>\n",
       "      <td>Pseudomonas</td>\n",
       "      <td>Pseudomonas_lactis</td>\n",
       "      <td>Carlino2024</td>\n",
       "      <td>AlvarezOrdonezA_xxxx__C02-19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AlvarezOrdonezA_xxxx__C02-19__bin.1</td>\n",
       "      <td>meat</td>\n",
       "      <td>sausage</td>\n",
       "      <td>meat</td>\n",
       "      <td>meat</td>\n",
       "      <td>fermented_meat</td>\n",
       "      <td>64.912281</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>712</td>\n",
       "      <td>3816144</td>\n",
       "      <td>...</td>\n",
       "      <td>6480</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Enterobacterales</td>\n",
       "      <td>Yersiniaceae</td>\n",
       "      <td>Ewingella</td>\n",
       "      <td>Ewingella_americana</td>\n",
       "      <td>Carlino2024</td>\n",
       "      <td>AlvarezOrdonezA_xxxx__C02-19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AlvarezOrdonezA_xxxx__C02-19__bin.10</td>\n",
       "      <td>meat</td>\n",
       "      <td>sausage</td>\n",
       "      <td>meat</td>\n",
       "      <td>meat</td>\n",
       "      <td>fermented_meat</td>\n",
       "      <td>63.900000</td>\n",
       "      <td>3.300000</td>\n",
       "      <td>1188</td>\n",
       "      <td>9316757</td>\n",
       "      <td>...</td>\n",
       "      <td>10532</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Ascomycota</td>\n",
       "      <td>Saccharomycetes</td>\n",
       "      <td>Saccharomycetales</td>\n",
       "      <td>Debaryomycetaceae</td>\n",
       "      <td>Debaryomyces</td>\n",
       "      <td>Debaryomyces_tyrocola</td>\n",
       "      <td>Carlino2024</td>\n",
       "      <td>AlvarezOrdonezA_xxxx__C02-19</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 21 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                 mag_id sample_description fermented_food  \\\n",
       "0   AlvarezOrdonezA_xxxx__C02-18__bin.2               meat        sausage   \n",
       "1   AlvarezOrdonezA_xxxx__C02-19__bin.5               meat        sausage   \n",
       "2   AlvarezOrdonezA_xxxx__C02-19__bin.3               meat        sausage   \n",
       "3   AlvarezOrdonezA_xxxx__C02-19__bin.1               meat        sausage   \n",
       "4  AlvarezOrdonezA_xxxx__C02-19__bin.10               meat        sausage   \n",
       "\n",
       "  specific_substrate substrate_category general_category  completeness  \\\n",
       "0               meat               meat   fermented_meat     51.600000   \n",
       "1               meat               meat   fermented_meat     63.793103   \n",
       "2               meat               meat   fermented_meat     74.137931   \n",
       "3               meat               meat   fermented_meat     64.912281   \n",
       "4               meat               meat   fermented_meat     63.900000   \n",
       "\n",
       "   contamination  contigs  total_length  ...    n50     domain  \\\n",
       "0       3.100000     2506      20998336  ...   9910  Eukaryota   \n",
       "1       0.862069      542       1596974  ...   3152   Bacteria   \n",
       "2       0.000000      279       5340756  ...  30953   Bacteria   \n",
       "3       0.000000      712       3816144  ...   6480   Bacteria   \n",
       "4       3.300000     1188       9316757  ...  10532  Eukaryota   \n",
       "\n",
       "           phylum                class              order             family  \\\n",
       "0      Ascomycota       Eurotiomycetes         Eurotiales     Aspergillaceae   \n",
       "1      Firmicutes              Bacilli         Bacillales  Staphylococcaceae   \n",
       "2  Proteobacteria  Gammaproteobacteria    Pseudomonadales   Pseudomonadaceae   \n",
       "3  Proteobacteria  Gammaproteobacteria   Enterobacterales       Yersiniaceae   \n",
       "4      Ascomycota      Saccharomycetes  Saccharomycetales  Debaryomycetaceae   \n",
       "\n",
       "            genus                 species study_catalog  \\\n",
       "0     Aspergillus     Aspergillus_glaucus   Carlino2024   \n",
       "1  Staphylococcus  Staphylococcus_equorum   Carlino2024   \n",
       "2     Pseudomonas      Pseudomonas_lactis   Carlino2024   \n",
       "3       Ewingella     Ewingella_americana   Carlino2024   \n",
       "4    Debaryomyces   Debaryomyces_tyrocola   Carlino2024   \n",
       "\n",
       "                         source  \n",
       "0  AlvarezOrdonezA_xxxx__C02-18  \n",
       "1  AlvarezOrdonezA_xxxx__C02-19  \n",
       "2  AlvarezOrdonezA_xxxx__C02-19  \n",
       "3  AlvarezOrdonezA_xxxx__C02-19  \n",
       "4  AlvarezOrdonezA_xxxx__C02-19  \n",
       "\n",
       "[5 rows x 21 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# loading Fermentos metadata file\n",
    "dataset = pd.read_csv('data/2025-03-24-all-ff-mag-metadata-cleaned-curated.tsv', sep='\\t')\n",
    "display(dataset.head())\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>dataset_name</th>\n",
       "      <th>sample_id</th>\n",
       "      <th>macrocategory</th>\n",
       "      <th>category</th>\n",
       "      <th>type</th>\n",
       "      <th>subtype</th>\n",
       "      <th>commercial_name</th>\n",
       "      <th>fermented/non-fermented</th>\n",
       "      <th>country</th>\n",
       "      <th>sample_accession</th>\n",
       "      <th>...</th>\n",
       "      <th>median_read_len</th>\n",
       "      <th>mean_read_len</th>\n",
       "      <th>max_read_len</th>\n",
       "      <th>n_contigs</th>\n",
       "      <th>n_MAGs_MQ_prok</th>\n",
       "      <th>n_MAGs_HQ_prok</th>\n",
       "      <th>n_MAGs_MQ_euk</th>\n",
       "      <th>n_MAGs_HQ_euk</th>\n",
       "      <th>filtered</th>\n",
       "      <th>curator</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AlvarezOrdonezA_xxxx</td>\n",
       "      <td>C01-16</td>\n",
       "      <td>food</td>\n",
       "      <td>meat</td>\n",
       "      <td>cured_meat_before_ripening</td>\n",
       "      <td>cured_meat_before_ripening</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NF</td>\n",
       "      <td>ESP</td>\n",
       "      <td>SAMN36689796</td>\n",
       "      <td>...</td>\n",
       "      <td>151.0</td>\n",
       "      <td>146.200505</td>\n",
       "      <td>151.0</td>\n",
       "      <td>62700</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>no</td>\n",
       "      <td>Niccolò_Carlino</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AlvarezOrdonezA_xxxx</td>\n",
       "      <td>C01-17</td>\n",
       "      <td>food</td>\n",
       "      <td>meat</td>\n",
       "      <td>cured_meat_before_ripening</td>\n",
       "      <td>cured_meat_before_ripening</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NF</td>\n",
       "      <td>ESP</td>\n",
       "      <td>SAMN36689797</td>\n",
       "      <td>...</td>\n",
       "      <td>151.0</td>\n",
       "      <td>147.266426</td>\n",
       "      <td>151.0</td>\n",
       "      <td>98756</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>no</td>\n",
       "      <td>Niccolò_Carlino</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AlvarezOrdonezA_xxxx</td>\n",
       "      <td>C01-18</td>\n",
       "      <td>food</td>\n",
       "      <td>fermented_meat</td>\n",
       "      <td>meat</td>\n",
       "      <td>cured_meat_after_ripening</td>\n",
       "      <td>NaN</td>\n",
       "      <td>F</td>\n",
       "      <td>ESP</td>\n",
       "      <td>SAMN36689824</td>\n",
       "      <td>...</td>\n",
       "      <td>151.0</td>\n",
       "      <td>147.760911</td>\n",
       "      <td>151.0</td>\n",
       "      <td>103911</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>no</td>\n",
       "      <td>Niccolò_Carlino</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AlvarezOrdonezA_xxxx</td>\n",
       "      <td>C01-19</td>\n",
       "      <td>food</td>\n",
       "      <td>fermented_meat</td>\n",
       "      <td>meat</td>\n",
       "      <td>cured_meat_after_ripening</td>\n",
       "      <td>NaN</td>\n",
       "      <td>F</td>\n",
       "      <td>ESP</td>\n",
       "      <td>SAMN36689825</td>\n",
       "      <td>...</td>\n",
       "      <td>151.0</td>\n",
       "      <td>149.200530</td>\n",
       "      <td>151.0</td>\n",
       "      <td>134121</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>no</td>\n",
       "      <td>Niccolò_Carlino</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AlvarezOrdonezA_xxxx</td>\n",
       "      <td>C02-16</td>\n",
       "      <td>food</td>\n",
       "      <td>meat</td>\n",
       "      <td>cured_meat_before_ripening</td>\n",
       "      <td>cured_meat_before_ripening</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NF</td>\n",
       "      <td>ESP</td>\n",
       "      <td>SAMN36689798</td>\n",
       "      <td>...</td>\n",
       "      <td>151.0</td>\n",
       "      <td>148.854534</td>\n",
       "      <td>151.0</td>\n",
       "      <td>37597</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>no</td>\n",
       "      <td>Niccolò_Carlino</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 32 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           dataset_name sample_id macrocategory        category  \\\n",
       "0  AlvarezOrdonezA_xxxx    C01-16          food            meat   \n",
       "1  AlvarezOrdonezA_xxxx    C01-17          food            meat   \n",
       "2  AlvarezOrdonezA_xxxx    C01-18          food  fermented_meat   \n",
       "3  AlvarezOrdonezA_xxxx    C01-19          food  fermented_meat   \n",
       "4  AlvarezOrdonezA_xxxx    C02-16          food            meat   \n",
       "\n",
       "                         type                     subtype commercial_name  \\\n",
       "0  cured_meat_before_ripening  cured_meat_before_ripening             NaN   \n",
       "1  cured_meat_before_ripening  cured_meat_before_ripening             NaN   \n",
       "2                        meat   cured_meat_after_ripening             NaN   \n",
       "3                        meat   cured_meat_after_ripening             NaN   \n",
       "4  cured_meat_before_ripening  cured_meat_before_ripening             NaN   \n",
       "\n",
       "  fermented/non-fermented country sample_accession  ... median_read_len  \\\n",
       "0                      NF     ESP     SAMN36689796  ...           151.0   \n",
       "1                      NF     ESP     SAMN36689797  ...           151.0   \n",
       "2                       F     ESP     SAMN36689824  ...           151.0   \n",
       "3                       F     ESP     SAMN36689825  ...           151.0   \n",
       "4                      NF     ESP     SAMN36689798  ...           151.0   \n",
       "\n",
       "  mean_read_len max_read_len n_contigs n_MAGs_MQ_prok n_MAGs_HQ_prok  \\\n",
       "0    146.200505        151.0     62700              0              0   \n",
       "1    147.266426        151.0     98756              0              0   \n",
       "2    147.760911        151.0    103911              0              0   \n",
       "3    149.200530        151.0    134121              0              0   \n",
       "4    148.854534        151.0     37597              1              1   \n",
       "\n",
       "  n_MAGs_MQ_euk n_MAGs_HQ_euk filtered          curator  \n",
       "0             0             0       no  Niccolò_Carlino  \n",
       "1             0             0       no  Niccolò_Carlino  \n",
       "2             0             0       no  Niccolò_Carlino  \n",
       "3             0             0       no  Niccolò_Carlino  \n",
       "4             0             0       no  Niccolò_Carlino  \n",
       "\n",
       "[5 rows x 32 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Dataset</th>\n",
       "      <th># Samples</th>\n",
       "      <th>Accession number</th>\n",
       "      <th>Reference</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AlvarezOrdonezA_xxxx</td>\n",
       "      <td>173</td>\n",
       "      <td>PRJNA997800, PRJNA997801</td>\n",
       "      <td>*</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ArikanM_2020</td>\n",
       "      <td>12</td>\n",
       "      <td>PRJEB22200</td>\n",
       "      <td>Arıkan, M., Mitchell, A.L., Finn, R.D., and Gü...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>BertuzziAS_2018</td>\n",
       "      <td>42</td>\n",
       "      <td>PRJEB15423</td>\n",
       "      <td>Bertuzzi, A.S., Walsh, A.M., Sheehan, J.J., Co...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ChaconVargasK_2020</td>\n",
       "      <td>5</td>\n",
       "      <td>PRJNA603591</td>\n",
       "      <td>Chacón-Vargas, K., Torres, J., Giles-Gómez, M....</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CM_INJERA</td>\n",
       "      <td>2</td>\n",
       "      <td>PRJNA504891</td>\n",
       "      <td>Manara, S., Selma-Royo, M., Huang, K.D., Asnic...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                Dataset  # Samples          Accession number  \\\n",
       "0  AlvarezOrdonezA_xxxx        173  PRJNA997800, PRJNA997801   \n",
       "1          ArikanM_2020         12                PRJEB22200   \n",
       "2       BertuzziAS_2018         42                PRJEB15423   \n",
       "3    ChaconVargasK_2020          5               PRJNA603591   \n",
       "4             CM_INJERA          2               PRJNA504891   \n",
       "\n",
       "                                           Reference  \n",
       "0                                                  *  \n",
       "1  Arıkan, M., Mitchell, A.L., Finn, R.D., and Gü...  \n",
       "2  Bertuzzi, A.S., Walsh, A.M., Sheehan, J.J., Co...  \n",
       "3  Chacón-Vargas, K., Torres, J., Giles-Gómez, M....  \n",
       "4  Manara, S., Selma-Royo, M., Huang, K.D., Asnic...  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# loading the Carlino metadata. \n",
    "carlino_metadata = pd.read_csv('data/cFMD_metadata.tsv', sep='\\t')\n",
    "carlino_supp = pd.read_excel('data/Carlino_supp.xlsx') #requires openpyxl to be installed\n",
    "\n",
    "display(carlino_metadata.head())\n",
    "display(carlino_supp.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total values after splitting: 16\n",
      "\n",
      "Rows with multiple accession numbers:\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Dataset</th>\n",
       "      <th># Samples</th>\n",
       "      <th>Accession number</th>\n",
       "      <th>Reference</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AlvarezOrdonezA_xxxx</td>\n",
       "      <td>173</td>\n",
       "      <td>PRJNA997800, PRJNA997801</td>\n",
       "      <td>Carlino, N., Blanco-Míguez, A., Punčochář, M.,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>CrovadoreJ_2017</td>\n",
       "      <td>5</td>\n",
       "      <td>PRJNA358537, PRJNA358558, PRJNA362533, PRJNA36...</td>\n",
       "      <td>Crovadore, J., Gérard, F., Chablais, R., Cocha...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>MASTER_WP4_CSIC_2</td>\n",
       "      <td>164</td>\n",
       "      <td>PRJNA997801, PRJNA1094419</td>\n",
       "      <td>Carlino, N., Blanco-Míguez, A., Punčochář, M.,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>MASTER_WP4_TEAGASC_1</td>\n",
       "      <td>156</td>\n",
       "      <td>PRGEB63604, PRJNA1094419</td>\n",
       "      <td>Carlino, N., Blanco-Míguez, A., Punčochář, M.,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>MASTER_WP4_UNINA_2</td>\n",
       "      <td>150</td>\n",
       "      <td>PRJNA897099, PRJNA997801, PRJNA1094419</td>\n",
       "      <td>Carlino, N., Blanco-Míguez, A., Punčochář, M.,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>MortensenS_xxxx</td>\n",
       "      <td>375</td>\n",
       "      <td>PRJNA977472, PRJNA1023503</td>\n",
       "      <td>Carlino, N., Blanco-Míguez, A., Punčochář, M.,...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 Dataset  # Samples  \\\n",
       "0   AlvarezOrdonezA_xxxx        173   \n",
       "5        CrovadoreJ_2017          5   \n",
       "26     MASTER_WP4_CSIC_2        164   \n",
       "31  MASTER_WP4_TEAGASC_1        156   \n",
       "33    MASTER_WP4_UNINA_2        150   \n",
       "36       MortensenS_xxxx        375   \n",
       "\n",
       "                                     Accession number  \\\n",
       "0                            PRJNA997800, PRJNA997801   \n",
       "5   PRJNA358537, PRJNA358558, PRJNA362533, PRJNA36...   \n",
       "26                          PRJNA997801, PRJNA1094419   \n",
       "31                           PRGEB63604, PRJNA1094419   \n",
       "33             PRJNA897099, PRJNA997801, PRJNA1094419   \n",
       "36                          PRJNA977472, PRJNA1023503   \n",
       "\n",
       "                                            Reference  \n",
       "0   Carlino, N., Blanco-Míguez, A., Punčochář, M.,...  \n",
       "5   Crovadore, J., Gérard, F., Chablais, R., Cocha...  \n",
       "26  Carlino, N., Blanco-Míguez, A., Punčochář, M.,...  \n",
       "31  Carlino, N., Blanco-Míguez, A., Punčochář, M.,...  \n",
       "33  Carlino, N., Blanco-Míguez, A., Punčochář, M.,...  \n",
       "36  Carlino, N., Blanco-Míguez, A., Punčochář, M.,...  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Number of rows with multiple accession numbers: 6\n",
      "total number of rows in dataset before splitting: 59\n",
      "number of rows at the end: 69\n"
     ]
    }
   ],
   "source": [
    "# organising carlino_supp\n",
    "    # replacing * with the Carlino paper\n",
    "    # spliting rows with multiple bioproject accession numbers\n",
    "\n",
    "# Replacing '*' in the Reference column with the full citation of the Carlino paper\n",
    "ref = 'Carlino, N., Blanco-Míguez, A., Punčochář, M., Mengoni, C., Pinto, F., Tatti, A., Manghi, P., Armanini, F., Avagliano, M., Barcenilla, C., Breselge, S., Cabrera-Rubio, R., Calvete-Torre, I., Coakley, M., Cobo-Díaz, J. F., De Filippis, F., Dey, H., Leech, J., Klaassens, E. S., Knobloch, S., O’Neil, D., Quijada, N. M., Sabater, C., Skírnisdóttir, S., Valentino, V., Walsh, L., MASTER EU Consortium, Asnicar, F., Fackelmann, G., Heidrich, V., Rota Stabelli, O., Pasolli, E., Arendt, E., Rubino, F., Cotter, P. D., Ruiz, L., Ercolini, D., Morrison, S., Marchart, J., Sessitsch, A., and Zannini, E. (2024). A pan-European study of the human gut microbiome reveals links with diet, health, and lifestyle. Cell, 187(13), 5795–5816.e15. https://doi.org/10.1016/j.cell.2024.05.028'\n",
    "carlino_supp.loc[carlino_supp['Reference'] == '*', 'Reference'] = ref\n",
    "\n",
    "# Splitting rows with multiple bioproject accession numbers\n",
    "# check how many rows have multiple Bioproject accession numbers\n",
    "# Find rows with commas in 'Accession number' and count values after splitting\n",
    "rows_with_commas = carlino_supp[carlino_supp['Accession number'].str.contains(',', na=False)]\n",
    "total_values = sum(len(accession.split(',')) for accession in rows_with_commas['Accession number'])\n",
    "print(f\"Total values after splitting: {total_values}\")\n",
    "\n",
    "print(\"\\nRows with multiple accession numbers:\")\n",
    "display(rows_with_commas)\n",
    "print(f\"\\nNumber of rows with multiple accession numbers: {rows_with_commas.shape[0]}\")\n",
    "print('total number of rows in dataset before splitting:', carlino_supp.shape[0])\n",
    "\n",
    "carlino_supp['Accession number'] = carlino_supp['Accession number'].str.split(',')\n",
    "carlino_supp = carlino_supp.explode('Accession number')\n",
    "# strip white spaces in the 'Accession number' column\n",
    "carlino_supp['Accession number'] = carlino_supp['Accession number'].str.strip()\n",
    "\n",
    "end_rows = carlino_supp.shape[0]\n",
    "print('number of rows at the end:', end_rows)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of rows with References: 2477\n",
      "Total rows in carlino_metadata: 3368\n"
     ]
    }
   ],
   "source": [
    "# Add Reference column from carlino_supp to carlino_metadata\n",
    "# Create a dictionary from carlino_supp, keeping the first occurrence of each project_accession\n",
    "reference_dict = dict(zip(carlino_supp['Accession number'], carlino_supp['Reference']))\n",
    "carlino_metadata['Reference'] = carlino_metadata['project_accession'].map(reference_dict)\n",
    "\n",
    "# Verify the merge by checking if References were added\n",
    "print(\"Number of rows with References:\", carlino_metadata['Reference'].notna().sum())\n",
    "print(\"Total rows in carlino_metadata:\", len(carlino_metadata))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>dataset_name</th>\n",
       "      <th>project_accession</th>\n",
       "      <th>Reference</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>408</th>\n",
       "      <td>KastmanEK_2016</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>799</th>\n",
       "      <td>MASTER_WP4_CSIC_2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2035</th>\n",
       "      <td>PorcellatoD_2016</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2469</th>\n",
       "      <td>WolfeBE_2014</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2493</th>\n",
       "      <td>XieM_2019</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2533</th>\n",
       "      <td>AlmeidaO_2020</td>\n",
       "      <td>PRJNA527768</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2547</th>\n",
       "      <td>CM_UNINA_FFOOD</td>\n",
       "      <td>PRJNA1206071</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2808</th>\n",
       "      <td>DecadtH_2024</td>\n",
       "      <td>PRJEB64331</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2827</th>\n",
       "      <td>FalardeauJ_2023</td>\n",
       "      <td>PRJNA863305</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2883</th>\n",
       "      <td>FontanaF_2023</td>\n",
       "      <td>PRJNA865096</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3011</th>\n",
       "      <td>FranciosaI_2021</td>\n",
       "      <td>PRJNA636619</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3026</th>\n",
       "      <td>GonzalezOrozcoB_2023</td>\n",
       "      <td>PRJNA843464</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3028</th>\n",
       "      <td>KharnaiorP_2023</td>\n",
       "      <td>PRJNA914873</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3032</th>\n",
       "      <td>LimaC_2020</td>\n",
       "      <td>PRJNA552479</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3046</th>\n",
       "      <td>LopezSanchezR_2023</td>\n",
       "      <td>PRJNA648868</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3050</th>\n",
       "      <td>MagliuloR_2024</td>\n",
       "      <td>PRJNA1084214</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3099</th>\n",
       "      <td>MotaGutierrezJ_2021</td>\n",
       "      <td>PRJNA627182</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3119</th>\n",
       "      <td>OlgaP_2019</td>\n",
       "      <td>PRJNA636820</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3121</th>\n",
       "      <td>OlgaP_2019</td>\n",
       "      <td>PRJNA636837</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3123</th>\n",
       "      <td>OlgaP_2019</td>\n",
       "      <td>PRJNA636891</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3125</th>\n",
       "      <td>OlgaP_2019</td>\n",
       "      <td>PRJNA637016</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3127</th>\n",
       "      <td>OlgaP_2019</td>\n",
       "      <td>PRJNA637018</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3128</th>\n",
       "      <td>QuijadaN_2022</td>\n",
       "      <td>PRJEB48589</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3130</th>\n",
       "      <td>SaakC_2023</td>\n",
       "      <td>PRJNA778418</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3148</th>\n",
       "      <td>SalgadoTS_2021</td>\n",
       "      <td>PRJNA704713</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3287</th>\n",
       "      <td>ShangpliangH_2023_a</td>\n",
       "      <td>PRJNA914731</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3289</th>\n",
       "      <td>ShangpliangH_2023_b</td>\n",
       "      <td>PRJNA914730</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3291</th>\n",
       "      <td>TomarS_2023</td>\n",
       "      <td>PRJNA763767</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3292</th>\n",
       "      <td>YangC_2021</td>\n",
       "      <td>PRJNA683931</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3318</th>\n",
       "      <td>YasirM_2022</td>\n",
       "      <td>PRJNA816670</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3336</th>\n",
       "      <td>YouL_2022</td>\n",
       "      <td>PRJNA687995</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3359</th>\n",
       "      <td>YuY_2022</td>\n",
       "      <td>PRJNA780248</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              dataset_name project_accession Reference\n",
       "408         KastmanEK_2016               NaN       NaN\n",
       "799      MASTER_WP4_CSIC_2               NaN       NaN\n",
       "2035      PorcellatoD_2016               NaN       NaN\n",
       "2469          WolfeBE_2014               NaN       NaN\n",
       "2493             XieM_2019               NaN       NaN\n",
       "2533         AlmeidaO_2020       PRJNA527768       NaN\n",
       "2547        CM_UNINA_FFOOD      PRJNA1206071       NaN\n",
       "2808          DecadtH_2024        PRJEB64331       NaN\n",
       "2827       FalardeauJ_2023       PRJNA863305       NaN\n",
       "2883         FontanaF_2023       PRJNA865096       NaN\n",
       "3011       FranciosaI_2021       PRJNA636619       NaN\n",
       "3026  GonzalezOrozcoB_2023       PRJNA843464       NaN\n",
       "3028       KharnaiorP_2023       PRJNA914873       NaN\n",
       "3032            LimaC_2020       PRJNA552479       NaN\n",
       "3046    LopezSanchezR_2023       PRJNA648868       NaN\n",
       "3050        MagliuloR_2024      PRJNA1084214       NaN\n",
       "3099   MotaGutierrezJ_2021       PRJNA627182       NaN\n",
       "3119            OlgaP_2019       PRJNA636820       NaN\n",
       "3121            OlgaP_2019       PRJNA636837       NaN\n",
       "3123            OlgaP_2019       PRJNA636891       NaN\n",
       "3125            OlgaP_2019       PRJNA637016       NaN\n",
       "3127            OlgaP_2019       PRJNA637018       NaN\n",
       "3128         QuijadaN_2022        PRJEB48589       NaN\n",
       "3130            SaakC_2023       PRJNA778418       NaN\n",
       "3148        SalgadoTS_2021       PRJNA704713       NaN\n",
       "3287   ShangpliangH_2023_a       PRJNA914731       NaN\n",
       "3289   ShangpliangH_2023_b       PRJNA914730       NaN\n",
       "3291           TomarS_2023       PRJNA763767       NaN\n",
       "3292            YangC_2021       PRJNA683931       NaN\n",
       "3318           YasirM_2022       PRJNA816670       NaN\n",
       "3336             YouL_2022       PRJNA687995       NaN\n",
       "3359              YuY_2022       PRJNA780248       NaN"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# checking the missing references\n",
    "missing_refs = carlino_metadata[\n",
    "    (carlino_metadata.Reference.isna()) & \n",
    "    (carlino_metadata['fermented/non-fermented'] == 'F')\n",
    "][['dataset_name', 'project_accession','Reference']].drop_duplicates()\n",
    "display(missing_refs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "# saving the missing references table to a file\n",
    "# will locate references manually\n",
    "missing_refs.to_csv('data/missing_references.tsv', sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>dataset_name</th>\n",
       "      <th># Samples</th>\n",
       "      <th>project_accession</th>\n",
       "      <th>Reference</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [dataset_name, # Samples, project_accession, Reference]\n",
       "Index: []"
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "carlino_supp[carlino_supp.project_accession == 'PRJNA1023503']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>dataset_name</th>\n",
       "      <th># Samples</th>\n",
       "      <th>project_accession</th>\n",
       "      <th>Reference</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>MortensenS_xxxx</td>\n",
       "      <td>375</td>\n",
       "      <td>PRJNA977472</td>\n",
       "      <td>Carlino, N., Blanco-Míguez, A., Punčochář, M.,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>MortensenS_xxxx</td>\n",
       "      <td>375</td>\n",
       "      <td>PRJNA1023503</td>\n",
       "      <td>Carlino, N., Blanco-Míguez, A., Punčochář, M.,...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       dataset_name  # Samples project_accession  \\\n",
       "36  MortensenS_xxxx        375       PRJNA977472   \n",
       "36  MortensenS_xxxx        375      PRJNA1023503   \n",
       "\n",
       "                                            Reference  \n",
       "36  Carlino, N., Blanco-Míguez, A., Punčochář, M.,...  \n",
       "36  Carlino, N., Blanco-Míguez, A., Punčochář, M.,...  "
      ]
     },
     "execution_count": 51,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "carlino_supp[carlino_supp.dataset_name == 'MortensenS_xxxx']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>dataset_name</th>\n",
       "      <th># Samples</th>\n",
       "      <th>project_accession</th>\n",
       "      <th>Reference</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [dataset_name, # Samples, project_accession, Reference]\n",
       "Index: []"
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "carlino_supp[carlino_supp.project_accession == 'PRJNA1023503']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['dataset_name', 'sample_id', 'macrocategory', 'category', 'type',\n",
       "       'subtype', 'commercial_name', 'fermented/non-fermented', 'country',\n",
       "       'sample_accession', 'run_accession', 'experiment_accession',\n",
       "       'study_accession', 'project_accession', 'database_origin',\n",
       "       'library_layout', 'sequencing_platform', 'DNA_extraction_kit',\n",
       "       'collection_date', 'n_of_bases', 'n_of_reads', 'min_read_len',\n",
       "       'median_read_len', 'mean_read_len', 'max_read_len', 'n_contigs',\n",
       "       'n_MAGs_MQ_prok', 'n_MAGs_HQ_prok', 'n_MAGs_MQ_euk', 'n_MAGs_HQ_euk',\n",
       "       'filtered', 'curator'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "carlino_metadata.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>mag_id</th>\n",
       "      <th>sample_description</th>\n",
       "      <th>fermented_food</th>\n",
       "      <th>specific_substrate</th>\n",
       "      <th>substrate_category</th>\n",
       "      <th>general_category</th>\n",
       "      <th>completeness</th>\n",
       "      <th>contamination</th>\n",
       "      <th>contigs</th>\n",
       "      <th>total_length</th>\n",
       "      <th>...</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>order</th>\n",
       "      <th>family</th>\n",
       "      <th>genus</th>\n",
       "      <th>species</th>\n",
       "      <th>study_catalog</th>\n",
       "      <th>source</th>\n",
       "      <th>dataset_name</th>\n",
       "      <th>sample_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AlvarezOrdonezA_xxxx__C02-18__bin.2</td>\n",
       "      <td>meat</td>\n",
       "      <td>sausage</td>\n",
       "      <td>meat</td>\n",
       "      <td>meat</td>\n",
       "      <td>fermented_meat</td>\n",
       "      <td>51.600000</td>\n",
       "      <td>3.100000</td>\n",
       "      <td>2506</td>\n",
       "      <td>20998336</td>\n",
       "      <td>...</td>\n",
       "      <td>Ascomycota</td>\n",
       "      <td>Eurotiomycetes</td>\n",
       "      <td>Eurotiales</td>\n",
       "      <td>Aspergillaceae</td>\n",
       "      <td>Aspergillus</td>\n",
       "      <td>Aspergillus_glaucus</td>\n",
       "      <td>Carlino2024</td>\n",
       "      <td>AlvarezOrdonezA_xxxx__C02-18</td>\n",
       "      <td>AlvarezOrdonezA_xxxx</td>\n",
       "      <td>C02-18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AlvarezOrdonezA_xxxx__C02-19__bin.5</td>\n",
       "      <td>meat</td>\n",
       "      <td>sausage</td>\n",
       "      <td>meat</td>\n",
       "      <td>meat</td>\n",
       "      <td>fermented_meat</td>\n",
       "      <td>63.793103</td>\n",
       "      <td>0.862069</td>\n",
       "      <td>542</td>\n",
       "      <td>1596974</td>\n",
       "      <td>...</td>\n",
       "      <td>Firmicutes</td>\n",
       "      <td>Bacilli</td>\n",
       "      <td>Bacillales</td>\n",
       "      <td>Staphylococcaceae</td>\n",
       "      <td>Staphylococcus</td>\n",
       "      <td>Staphylococcus_equorum</td>\n",
       "      <td>Carlino2024</td>\n",
       "      <td>AlvarezOrdonezA_xxxx__C02-19</td>\n",
       "      <td>AlvarezOrdonezA_xxxx</td>\n",
       "      <td>C02-19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AlvarezOrdonezA_xxxx__C02-19__bin.3</td>\n",
       "      <td>meat</td>\n",
       "      <td>sausage</td>\n",
       "      <td>meat</td>\n",
       "      <td>meat</td>\n",
       "      <td>fermented_meat</td>\n",
       "      <td>74.137931</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>279</td>\n",
       "      <td>5340756</td>\n",
       "      <td>...</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Pseudomonadales</td>\n",
       "      <td>Pseudomonadaceae</td>\n",
       "      <td>Pseudomonas</td>\n",
       "      <td>Pseudomonas_lactis</td>\n",
       "      <td>Carlino2024</td>\n",
       "      <td>AlvarezOrdonezA_xxxx__C02-19</td>\n",
       "      <td>AlvarezOrdonezA_xxxx</td>\n",
       "      <td>C02-19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AlvarezOrdonezA_xxxx__C02-19__bin.1</td>\n",
       "      <td>meat</td>\n",
       "      <td>sausage</td>\n",
       "      <td>meat</td>\n",
       "      <td>meat</td>\n",
       "      <td>fermented_meat</td>\n",
       "      <td>64.912281</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>712</td>\n",
       "      <td>3816144</td>\n",
       "      <td>...</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Enterobacterales</td>\n",
       "      <td>Yersiniaceae</td>\n",
       "      <td>Ewingella</td>\n",
       "      <td>Ewingella_americana</td>\n",
       "      <td>Carlino2024</td>\n",
       "      <td>AlvarezOrdonezA_xxxx__C02-19</td>\n",
       "      <td>AlvarezOrdonezA_xxxx</td>\n",
       "      <td>C02-19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AlvarezOrdonezA_xxxx__C02-19__bin.10</td>\n",
       "      <td>meat</td>\n",
       "      <td>sausage</td>\n",
       "      <td>meat</td>\n",
       "      <td>meat</td>\n",
       "      <td>fermented_meat</td>\n",
       "      <td>63.900000</td>\n",
       "      <td>3.300000</td>\n",
       "      <td>1188</td>\n",
       "      <td>9316757</td>\n",
       "      <td>...</td>\n",
       "      <td>Ascomycota</td>\n",
       "      <td>Saccharomycetes</td>\n",
       "      <td>Saccharomycetales</td>\n",
       "      <td>Debaryomycetaceae</td>\n",
       "      <td>Debaryomyces</td>\n",
       "      <td>Debaryomyces_tyrocola</td>\n",
       "      <td>Carlino2024</td>\n",
       "      <td>AlvarezOrdonezA_xxxx__C02-19</td>\n",
       "      <td>AlvarezOrdonezA_xxxx</td>\n",
       "      <td>C02-19</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 23 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                 mag_id sample_description fermented_food  \\\n",
       "0   AlvarezOrdonezA_xxxx__C02-18__bin.2               meat        sausage   \n",
       "1   AlvarezOrdonezA_xxxx__C02-19__bin.5               meat        sausage   \n",
       "2   AlvarezOrdonezA_xxxx__C02-19__bin.3               meat        sausage   \n",
       "3   AlvarezOrdonezA_xxxx__C02-19__bin.1               meat        sausage   \n",
       "4  AlvarezOrdonezA_xxxx__C02-19__bin.10               meat        sausage   \n",
       "\n",
       "  specific_substrate substrate_category general_category  completeness  \\\n",
       "0               meat               meat   fermented_meat     51.600000   \n",
       "1               meat               meat   fermented_meat     63.793103   \n",
       "2               meat               meat   fermented_meat     74.137931   \n",
       "3               meat               meat   fermented_meat     64.912281   \n",
       "4               meat               meat   fermented_meat     63.900000   \n",
       "\n",
       "   contamination  contigs  total_length  ...          phylum  \\\n",
       "0       3.100000     2506      20998336  ...      Ascomycota   \n",
       "1       0.862069      542       1596974  ...      Firmicutes   \n",
       "2       0.000000      279       5340756  ...  Proteobacteria   \n",
       "3       0.000000      712       3816144  ...  Proteobacteria   \n",
       "4       3.300000     1188       9316757  ...      Ascomycota   \n",
       "\n",
       "                 class              order             family           genus  \\\n",
       "0       Eurotiomycetes         Eurotiales     Aspergillaceae     Aspergillus   \n",
       "1              Bacilli         Bacillales  Staphylococcaceae  Staphylococcus   \n",
       "2  Gammaproteobacteria    Pseudomonadales   Pseudomonadaceae     Pseudomonas   \n",
       "3  Gammaproteobacteria   Enterobacterales       Yersiniaceae       Ewingella   \n",
       "4      Saccharomycetes  Saccharomycetales  Debaryomycetaceae    Debaryomyces   \n",
       "\n",
       "                  species study_catalog                        source  \\\n",
       "0     Aspergillus_glaucus   Carlino2024  AlvarezOrdonezA_xxxx__C02-18   \n",
       "1  Staphylococcus_equorum   Carlino2024  AlvarezOrdonezA_xxxx__C02-19   \n",
       "2      Pseudomonas_lactis   Carlino2024  AlvarezOrdonezA_xxxx__C02-19   \n",
       "3     Ewingella_americana   Carlino2024  AlvarezOrdonezA_xxxx__C02-19   \n",
       "4   Debaryomyces_tyrocola   Carlino2024  AlvarezOrdonezA_xxxx__C02-19   \n",
       "\n",
       "           dataset_name sample_id  \n",
       "0  AlvarezOrdonezA_xxxx    C02-18  \n",
       "1  AlvarezOrdonezA_xxxx    C02-19  \n",
       "2  AlvarezOrdonezA_xxxx    C02-19  \n",
       "3  AlvarezOrdonezA_xxxx    C02-19  \n",
       "4  AlvarezOrdonezA_xxxx    C02-19  \n",
       "\n",
       "[5 rows x 23 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Split source column into dataset_name and sample_id\n",
    "dataset[['dataset_name', 'sample_id']] = dataset['mag_id'].str.rsplit('__', n=2, expand=True).iloc[:, :2]\n",
    "dataset.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(13858, 23)"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dataset.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Found 2 duplicate entries\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>dataset_name</th>\n",
       "      <th>sample_id</th>\n",
       "      <th>count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>YouL_2022</td>\n",
       "      <td>SRR13308535</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  dataset_name    sample_id  count\n",
       "0    YouL_2022  SRR13308535      2"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Duplicate rows:\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>dataset_name</th>\n",
       "      <th>sample_id</th>\n",
       "      <th>macrocategory</th>\n",
       "      <th>category</th>\n",
       "      <th>type</th>\n",
       "      <th>subtype</th>\n",
       "      <th>commercial_name</th>\n",
       "      <th>fermented/non-fermented</th>\n",
       "      <th>country</th>\n",
       "      <th>sample_accession</th>\n",
       "      <th>...</th>\n",
       "      <th>median_read_len</th>\n",
       "      <th>mean_read_len</th>\n",
       "      <th>max_read_len</th>\n",
       "      <th>n_contigs</th>\n",
       "      <th>n_MAGs_MQ_prok</th>\n",
       "      <th>n_MAGs_HQ_prok</th>\n",
       "      <th>n_MAGs_MQ_euk</th>\n",
       "      <th>n_MAGs_HQ_euk</th>\n",
       "      <th>filtered</th>\n",
       "      <th>curator</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3351</th>\n",
       "      <td>YouL_2022</td>\n",
       "      <td>SRR13308535</td>\n",
       "      <td>food</td>\n",
       "      <td>dairy</td>\n",
       "      <td>koumiss</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>F</td>\n",
       "      <td>CHN</td>\n",
       "      <td>SAMN17165130</td>\n",
       "      <td>...</td>\n",
       "      <td>150.0</td>\n",
       "      <td>149.562479</td>\n",
       "      <td>150.0</td>\n",
       "      <td>10491</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>no</td>\n",
       "      <td>Hrituraj_Dey</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3352</th>\n",
       "      <td>YouL_2022</td>\n",
       "      <td>SRR13308535</td>\n",
       "      <td>food</td>\n",
       "      <td>dairy</td>\n",
       "      <td>koumiss</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>F</td>\n",
       "      <td>CHN</td>\n",
       "      <td>SAMN17165130</td>\n",
       "      <td>...</td>\n",
       "      <td>150.0</td>\n",
       "      <td>149.562479</td>\n",
       "      <td>150.0</td>\n",
       "      <td>10491</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>no</td>\n",
       "      <td>Hrituraj_Dey</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2 rows × 32 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     dataset_name    sample_id macrocategory category     type subtype  \\\n",
       "3351    YouL_2022  SRR13308535          food    dairy  koumiss     NaN   \n",
       "3352    YouL_2022  SRR13308535          food    dairy  koumiss     NaN   \n",
       "\n",
       "     commercial_name fermented/non-fermented country sample_accession  ...  \\\n",
       "3351             NaN                       F     CHN     SAMN17165130  ...   \n",
       "3352             NaN                       F     CHN     SAMN17165130  ...   \n",
       "\n",
       "     median_read_len mean_read_len max_read_len n_contigs n_MAGs_MQ_prok  \\\n",
       "3351           150.0    149.562479        150.0     10491              1   \n",
       "3352           150.0    149.562479        150.0     10491              1   \n",
       "\n",
       "     n_MAGs_HQ_prok n_MAGs_MQ_euk n_MAGs_HQ_euk filtered       curator  \n",
       "3351              0             1             1       no  Hrituraj_Dey  \n",
       "3352              0             1             1       no  Hrituraj_Dey  \n",
       "\n",
       "[2 rows x 32 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Check for duplicates in carlino_metadata using dataset_name and sample_id as keys\n",
    "duplicates = carlino_metadata[carlino_metadata.duplicated(['dataset_name', 'sample_id'], keep=False)]\n",
    "\n",
    "# Display counts of duplicates if any exist\n",
    "if len(duplicates) > 0:\n",
    "    print(f\"Found {len(duplicates)} duplicate entries\")\n",
    "    # Group duplicates and show their counts\n",
    "    duplicate_counts = duplicates.groupby(['dataset_name', 'sample_id']).size().reset_index(name='count')\n",
    "    display(duplicate_counts)\n",
    "    \n",
    "    # Display the actual duplicate rows\n",
    "    print(\"\\nDuplicate rows:\")\n",
    "    display(duplicates)\n",
    "else:\n",
    "    print(\"No duplicates found in carlino_metadata\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of duplicates after removal: 0\n"
     ]
    }
   ],
   "source": [
    "# Drop duplicates keeping the first occurrence\n",
    "carlino_metadata = carlino_metadata.drop_duplicates(['sample_id'], keep='first')\n",
    "\n",
    "# Verify the duplicate is removed\n",
    "print(\"Number of duplicates after removal:\",\n",
    "    carlino_metadata.duplicated(['dataset_name', 'sample_id']).sum())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(3366, 32)"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "dataset_name                 85\n",
       "sample_id                  3366\n",
       "macrocategory                 1\n",
       "category                     15\n",
       "type                        118\n",
       "subtype                     337\n",
       "commercial_name              13\n",
       "fermented/non-fermented       2\n",
       "country                      52\n",
       "sample_accession           3303\n",
       "run_accession              3196\n",
       "experiment_accession       1367\n",
       "study_accession              85\n",
       "project_accession            85\n",
       "database_origin               4\n",
       "library_layout                3\n",
       "sequencing_platform          16\n",
       "DNA_extraction_kit           35\n",
       "collection_date             201\n",
       "n_of_bases                 3366\n",
       "n_of_reads                 3364\n",
       "min_read_len                 33\n",
       "median_read_len              62\n",
       "mean_read_len              3027\n",
       "max_read_len                 26\n",
       "n_contigs                  3065\n",
       "n_MAGs_MQ_prok               17\n",
       "n_MAGs_HQ_prok               18\n",
       "n_MAGs_MQ_euk                 7\n",
       "n_MAGs_HQ_euk                 6\n",
       "filtered                      2\n",
       "curator                       4\n",
       "dtype: int64"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# check the number of unique values in each column of carlino_metadata\n",
    "display(carlino_metadata.shape, carlino_metadata.nunique())\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Shape after merge:\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "(13858, 38)"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "(13858, 23)"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Merge dataset with selected columns from carlino_metadata\n",
    "merged_df = pd.merge(dataset, \n",
    "                    carlino_metadata.iloc[:, :17], \n",
    "                    how='left',\n",
    "                    on=['dataset_name', 'sample_id'])\n",
    "\n",
    "# Verify the merge\n",
    "print(\"\\nShape after merge:\")\n",
    "display(merged_df.shape, dataset.shape)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>count</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>study_catalog</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Du2023</th>\n",
       "      <td>553</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Caffrey2024</th>\n",
       "      <td>350</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Saak2023_CheeseMAGs</th>\n",
       "      <td>52</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Rappaport2024_sourdough</th>\n",
       "      <td>29</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Carlino2024</th>\n",
       "      <td>23</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                         count\n",
       "study_catalog                 \n",
       "Du2023                     553\n",
       "Caffrey2024                350\n",
       "Saak2023_CheeseMAGs         52\n",
       "Rappaport2024_sourdough     29\n",
       "Carlino2024                 23"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Total number of unique study_catalog values for samples without sequencing platform: 5\n",
      "\n",
      "Total samples without sequencing platform: 1007\n"
     ]
    }
   ],
   "source": [
    "#print the unique values of columns study_catalog for data that was not found in carlino\n",
    "# Get value counts of study_catalog for entries with missing sequencing_platform\n",
    "study_counts = merged_df[merged_df['sequencing_platform'].isna()]['study_catalog'].value_counts().to_frame()\n",
    "display(study_counts)\n",
    "print(f\"\\nTotal number of unique study_catalog values for samples without sequencing platform: {len(study_counts)}\")\n",
    "\n",
    "# Print total number of rows with missing sequencing_platform\n",
    "print(f\"\\nTotal samples without sequencing platform: {merged_df['sequencing_platform'].isna().sum()}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
