{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Data wrangling and curation of Fermentos dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd # type: ignore"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>mag_id</th>\n",
       "      <th>sample_description</th>\n",
       "      <th>fermented_food</th>\n",
       "      <th>specific_substrate</th>\n",
       "      <th>substrate_category</th>\n",
       "      <th>general_category</th>\n",
       "      <th>completeness</th>\n",
       "      <th>contamination</th>\n",
       "      <th>contigs</th>\n",
       "      <th>total_length</th>\n",
       "      <th>...</th>\n",
       "      <th>n50</th>\n",
       "      <th>domain</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>order</th>\n",
       "      <th>family</th>\n",
       "      <th>genus</th>\n",
       "      <th>species</th>\n",
       "      <th>study_catalog</th>\n",
       "      <th>source</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AlvarezOrdonezA_xxxx__C02-18__bin.2</td>\n",
       "      <td>meat</td>\n",
       "      <td>sausage</td>\n",
       "      <td>meat</td>\n",
       "      <td>meat</td>\n",
       "      <td>fermented_meat</td>\n",
       "      <td>51.600000</td>\n",
       "      <td>3.100000</td>\n",
       "      <td>2506</td>\n",
       "      <td>20998336</td>\n",
       "      <td>...</td>\n",
       "      <td>9910</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Ascomycota</td>\n",
       "      <td>Eurotiomycetes</td>\n",
       "      <td>Eurotiales</td>\n",
       "      <td>Aspergillaceae</td>\n",
       "      <td>Aspergillus</td>\n",
       "      <td>Aspergillus_glaucus</td>\n",
       "      <td>Carlino2024</td>\n",
       "      <td>AlvarezOrdonezA_xxxx__C02-18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AlvarezOrdonezA_xxxx__C02-19__bin.5</td>\n",
       "      <td>meat</td>\n",
       "      <td>sausage</td>\n",
       "      <td>meat</td>\n",
       "      <td>meat</td>\n",
       "      <td>fermented_meat</td>\n",
       "      <td>63.793103</td>\n",
       "      <td>0.862069</td>\n",
       "      <td>542</td>\n",
       "      <td>1596974</td>\n",
       "      <td>...</td>\n",
       "      <td>3152</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Firmicutes</td>\n",
       "      <td>Bacilli</td>\n",
       "      <td>Bacillales</td>\n",
       "      <td>Staphylococcaceae</td>\n",
       "      <td>Staphylococcus</td>\n",
       "      <td>Staphylococcus_equorum</td>\n",
       "      <td>Carlino2024</td>\n",
       "      <td>AlvarezOrdonezA_xxxx__C02-19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AlvarezOrdonezA_xxxx__C02-19__bin.3</td>\n",
       "      <td>meat</td>\n",
       "      <td>sausage</td>\n",
       "      <td>meat</td>\n",
       "      <td>meat</td>\n",
       "      <td>fermented_meat</td>\n",
       "      <td>74.137931</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>279</td>\n",
       "      <td>5340756</td>\n",
       "      <td>...</td>\n",
       "      <td>30953</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Pseudomonadales</td>\n",
       "      <td>Pseudomonadaceae</td>\n",
       "      <td>Pseudomonas</td>\n",
       "      <td>Pseudomonas_lactis</td>\n",
       "      <td>Carlino2024</td>\n",
       "      <td>AlvarezOrdonezA_xxxx__C02-19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AlvarezOrdonezA_xxxx__C02-19__bin.1</td>\n",
       "      <td>meat</td>\n",
       "      <td>sausage</td>\n",
       "      <td>meat</td>\n",
       "      <td>meat</td>\n",
       "      <td>fermented_meat</td>\n",
       "      <td>64.912281</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>712</td>\n",
       "      <td>3816144</td>\n",
       "      <td>...</td>\n",
       "      <td>6480</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Enterobacterales</td>\n",
       "      <td>Yersiniaceae</td>\n",
       "      <td>Ewingella</td>\n",
       "      <td>Ewingella_americana</td>\n",
       "      <td>Carlino2024</td>\n",
       "      <td>AlvarezOrdonezA_xxxx__C02-19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AlvarezOrdonezA_xxxx__C02-19__bin.10</td>\n",
       "      <td>meat</td>\n",
       "      <td>sausage</td>\n",
       "      <td>meat</td>\n",
       "      <td>meat</td>\n",
       "      <td>fermented_meat</td>\n",
       "      <td>63.900000</td>\n",
       "      <td>3.300000</td>\n",
       "      <td>1188</td>\n",
       "      <td>9316757</td>\n",
       "      <td>...</td>\n",
       "      <td>10532</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Ascomycota</td>\n",
       "      <td>Saccharomycetes</td>\n",
       "      <td>Saccharomycetales</td>\n",
       "      <td>Debaryomycetaceae</td>\n",
       "      <td>Debaryomyces</td>\n",
       "      <td>Debaryomyces_tyrocola</td>\n",
       "      <td>Carlino2024</td>\n",
       "      <td>AlvarezOrdonezA_xxxx__C02-19</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 21 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                 mag_id sample_description fermented_food  \\\n",
       "0   AlvarezOrdonezA_xxxx__C02-18__bin.2               meat        sausage   \n",
       "1   AlvarezOrdonezA_xxxx__C02-19__bin.5               meat        sausage   \n",
       "2   AlvarezOrdonezA_xxxx__C02-19__bin.3               meat        sausage   \n",
       "3   AlvarezOrdonezA_xxxx__C02-19__bin.1               meat        sausage   \n",
       "4  AlvarezOrdonezA_xxxx__C02-19__bin.10               meat        sausage   \n",
       "\n",
       "  specific_substrate substrate_category general_category  completeness  \\\n",
       "0               meat               meat   fermented_meat     51.600000   \n",
       "1               meat               meat   fermented_meat     63.793103   \n",
       "2               meat               meat   fermented_meat     74.137931   \n",
       "3               meat               meat   fermented_meat     64.912281   \n",
       "4               meat               meat   fermented_meat     63.900000   \n",
       "\n",
       "   contamination  contigs  total_length  ...    n50     domain  \\\n",
       "0       3.100000     2506      20998336  ...   9910  Eukaryota   \n",
       "1       0.862069      542       1596974  ...   3152   Bacteria   \n",
       "2       0.000000      279       5340756  ...  30953   Bacteria   \n",
       "3       0.000000      712       3816144  ...   6480   Bacteria   \n",
       "4       3.300000     1188       9316757  ...  10532  Eukaryota   \n",
       "\n",
       "           phylum                class              order             family  \\\n",
       "0      Ascomycota       Eurotiomycetes         Eurotiales     Aspergillaceae   \n",
       "1      Firmicutes              Bacilli         Bacillales  Staphylococcaceae   \n",
       "2  Proteobacteria  Gammaproteobacteria    Pseudomonadales   Pseudomonadaceae   \n",
       "3  Proteobacteria  Gammaproteobacteria   Enterobacterales       Yersiniaceae   \n",
       "4      Ascomycota      Saccharomycetes  Saccharomycetales  Debaryomycetaceae   \n",
       "\n",
       "            genus                 species study_catalog  \\\n",
       "0     Aspergillus     Aspergillus_glaucus   Carlino2024   \n",
       "1  Staphylococcus  Staphylococcus_equorum   Carlino2024   \n",
       "2     Pseudomonas      Pseudomonas_lactis   Carlino2024   \n",
       "3       Ewingella     Ewingella_americana   Carlino2024   \n",
       "4    Debaryomyces   Debaryomyces_tyrocola   Carlino2024   \n",
       "\n",
       "                         source  \n",
       "0  AlvarezOrdonezA_xxxx__C02-18  \n",
       "1  AlvarezOrdonezA_xxxx__C02-19  \n",
       "2  AlvarezOrdonezA_xxxx__C02-19  \n",
       "3  AlvarezOrdonezA_xxxx__C02-19  \n",
       "4  AlvarezOrdonezA_xxxx__C02-19  \n",
       "\n",
       "[5 rows x 21 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# loading Fermentos metadata file\n",
    "dataset = pd.read_csv('data/2025-03-24-all-ff-mag-metadata-cleaned-curated.tsv', sep='\\t')\n",
    "display(dataset.head())\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>dataset_name</th>\n",
       "      <th>sample_id</th>\n",
       "      <th>macrocategory</th>\n",
       "      <th>category</th>\n",
       "      <th>type</th>\n",
       "      <th>subtype</th>\n",
       "      <th>commercial_name</th>\n",
       "      <th>fermented/non-fermented</th>\n",
       "      <th>country</th>\n",
       "      <th>sample_accession</th>\n",
       "      <th>...</th>\n",
       "      <th>median_read_len</th>\n",
       "      <th>mean_read_len</th>\n",
       "      <th>max_read_len</th>\n",
       "      <th>n_contigs</th>\n",
       "      <th>n_MAGs_MQ_prok</th>\n",
       "      <th>n_MAGs_HQ_prok</th>\n",
       "      <th>n_MAGs_MQ_euk</th>\n",
       "      <th>n_MAGs_HQ_euk</th>\n",
       "      <th>filtered</th>\n",
       "      <th>curator</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AlvarezOrdonezA_xxxx</td>\n",
       "      <td>C01-16</td>\n",
       "      <td>food</td>\n",
       "      <td>meat</td>\n",
       "      <td>cured_meat_before_ripening</td>\n",
       "      <td>cured_meat_before_ripening</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NF</td>\n",
       "      <td>ESP</td>\n",
       "      <td>SAMN36689796</td>\n",
       "      <td>...</td>\n",
       "      <td>151.0</td>\n",
       "      <td>146.200505</td>\n",
       "      <td>151.0</td>\n",
       "      <td>62700</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>no</td>\n",
       "      <td>Niccolò_Carlino</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AlvarezOrdonezA_xxxx</td>\n",
       "      <td>C01-17</td>\n",
       "      <td>food</td>\n",
       "      <td>meat</td>\n",
       "      <td>cured_meat_before_ripening</td>\n",
       "      <td>cured_meat_before_ripening</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NF</td>\n",
       "      <td>ESP</td>\n",
       "      <td>SAMN36689797</td>\n",
       "      <td>...</td>\n",
       "      <td>151.0</td>\n",
       "      <td>147.266426</td>\n",
       "      <td>151.0</td>\n",
       "      <td>98756</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>no</td>\n",
       "      <td>Niccolò_Carlino</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AlvarezOrdonezA_xxxx</td>\n",
       "      <td>C01-18</td>\n",
       "      <td>food</td>\n",
       "      <td>fermented_meat</td>\n",
       "      <td>meat</td>\n",
       "      <td>cured_meat_after_ripening</td>\n",
       "      <td>NaN</td>\n",
       "      <td>F</td>\n",
       "      <td>ESP</td>\n",
       "      <td>SAMN36689824</td>\n",
       "      <td>...</td>\n",
       "      <td>151.0</td>\n",
       "      <td>147.760911</td>\n",
       "      <td>151.0</td>\n",
       "      <td>103911</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>no</td>\n",
       "      <td>Niccolò_Carlino</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AlvarezOrdonezA_xxxx</td>\n",
       "      <td>C01-19</td>\n",
       "      <td>food</td>\n",
       "      <td>fermented_meat</td>\n",
       "      <td>meat</td>\n",
       "      <td>cured_meat_after_ripening</td>\n",
       "      <td>NaN</td>\n",
       "      <td>F</td>\n",
       "      <td>ESP</td>\n",
       "      <td>SAMN36689825</td>\n",
       "      <td>...</td>\n",
       "      <td>151.0</td>\n",
       "      <td>149.200530</td>\n",
       "      <td>151.0</td>\n",
       "      <td>134121</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>no</td>\n",
       "      <td>Niccolò_Carlino</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AlvarezOrdonezA_xxxx</td>\n",
       "      <td>C02-16</td>\n",
       "      <td>food</td>\n",
       "      <td>meat</td>\n",
       "      <td>cured_meat_before_ripening</td>\n",
       "      <td>cured_meat_before_ripening</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NF</td>\n",
       "      <td>ESP</td>\n",
       "      <td>SAMN36689798</td>\n",
       "      <td>...</td>\n",
       "      <td>151.0</td>\n",
       "      <td>148.854534</td>\n",
       "      <td>151.0</td>\n",
       "      <td>37597</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>no</td>\n",
       "      <td>Niccolò_Carlino</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 32 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           dataset_name sample_id macrocategory        category  \\\n",
       "0  AlvarezOrdonezA_xxxx    C01-16          food            meat   \n",
       "1  AlvarezOrdonezA_xxxx    C01-17          food            meat   \n",
       "2  AlvarezOrdonezA_xxxx    C01-18          food  fermented_meat   \n",
       "3  AlvarezOrdonezA_xxxx    C01-19          food  fermented_meat   \n",
       "4  AlvarezOrdonezA_xxxx    C02-16          food            meat   \n",
       "\n",
       "                         type                     subtype commercial_name  \\\n",
       "0  cured_meat_before_ripening  cured_meat_before_ripening             NaN   \n",
       "1  cured_meat_before_ripening  cured_meat_before_ripening             NaN   \n",
       "2                        meat   cured_meat_after_ripening             NaN   \n",
       "3                        meat   cured_meat_after_ripening             NaN   \n",
       "4  cured_meat_before_ripening  cured_meat_before_ripening             NaN   \n",
       "\n",
       "  fermented/non-fermented country sample_accession  ... median_read_len  \\\n",
       "0                      NF     ESP     SAMN36689796  ...           151.0   \n",
       "1                      NF     ESP     SAMN36689797  ...           151.0   \n",
       "2                       F     ESP     SAMN36689824  ...           151.0   \n",
       "3                       F     ESP     SAMN36689825  ...           151.0   \n",
       "4                      NF     ESP     SAMN36689798  ...           151.0   \n",
       "\n",
       "  mean_read_len max_read_len n_contigs n_MAGs_MQ_prok n_MAGs_HQ_prok  \\\n",
       "0    146.200505        151.0     62700              0              0   \n",
       "1    147.266426        151.0     98756              0              0   \n",
       "2    147.760911        151.0    103911              0              0   \n",
       "3    149.200530        151.0    134121              0              0   \n",
       "4    148.854534        151.0     37597              1              1   \n",
       "\n",
       "  n_MAGs_MQ_euk n_MAGs_HQ_euk filtered          curator  \n",
       "0             0             0       no  Niccolò_Carlino  \n",
       "1             0             0       no  Niccolò_Carlino  \n",
       "2             0             0       no  Niccolò_Carlino  \n",
       "3             0             0       no  Niccolò_Carlino  \n",
       "4             0             0       no  Niccolò_Carlino  \n",
       "\n",
       "[5 rows x 32 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Dataset</th>\n",
       "      <th># Samples</th>\n",
       "      <th>Accession number</th>\n",
       "      <th>Reference</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AlvarezOrdonezA_xxxx</td>\n",
       "      <td>173</td>\n",
       "      <td>PRJNA997800, PRJNA997801</td>\n",
       "      <td>*</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ArikanM_2020</td>\n",
       "      <td>12</td>\n",
       "      <td>PRJEB22200</td>\n",
       "      <td>Arıkan, M., Mitchell, A.L., Finn, R.D., and Gü...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>BertuzziAS_2018</td>\n",
       "      <td>42</td>\n",
       "      <td>PRJEB15423</td>\n",
       "      <td>Bertuzzi, A.S., Walsh, A.M., Sheehan, J.J., Co...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ChaconVargasK_2020</td>\n",
       "      <td>5</td>\n",
       "      <td>PRJNA603591</td>\n",
       "      <td>Chacón-Vargas, K., Torres, J., Giles-Gómez, M....</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CM_INJERA</td>\n",
       "      <td>2</td>\n",
       "      <td>PRJNA504891</td>\n",
       "      <td>Manara, S., Selma-Royo, M., Huang, K.D., Asnic...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                Dataset  # Samples          Accession number  \\\n",
       "0  AlvarezOrdonezA_xxxx        173  PRJNA997800, PRJNA997801   \n",
       "1          ArikanM_2020         12                PRJEB22200   \n",
       "2       BertuzziAS_2018         42                PRJEB15423   \n",
       "3    ChaconVargasK_2020          5               PRJNA603591   \n",
       "4             CM_INJERA          2               PRJNA504891   \n",
       "\n",
       "                                           Reference  \n",
       "0                                                  *  \n",
       "1  Arıkan, M., Mitchell, A.L., Finn, R.D., and Gü...  \n",
       "2  Bertuzzi, A.S., Walsh, A.M., Sheehan, J.J., Co...  \n",
       "3  Chacón-Vargas, K., Torres, J., Giles-Gómez, M....  \n",
       "4  Manara, S., Selma-Royo, M., Huang, K.D., Asnic...  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# loading the Carlino metadata. \n",
    "carlino_metadata = pd.read_csv('data/cFMD_metadata.tsv', sep='\\t')\n",
    "carlino_supp = pd.read_excel('data/Carlino_supp.xlsx') #requires openpyxl to be installed\n",
    "\n",
    "display(carlino_metadata.head())\n",
    "display(carlino_supp.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total values after splitting: 16\n",
      "\n",
      "Rows with multiple accession numbers:\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Dataset</th>\n",
       "      <th># Samples</th>\n",
       "      <th>Accession number</th>\n",
       "      <th>Reference</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AlvarezOrdonezA_xxxx</td>\n",
       "      <td>173</td>\n",
       "      <td>PRJNA997800, PRJNA997801</td>\n",
       "      <td>Carlino, N., Blanco-Míguez, A., Punčochář, M.,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>CrovadoreJ_2017</td>\n",
       "      <td>5</td>\n",
       "      <td>PRJNA358537, PRJNA358558, PRJNA362533, PRJNA36...</td>\n",
       "      <td>Crovadore, J., Gérard, F., Chablais, R., Cocha...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>MASTER_WP4_CSIC_2</td>\n",
       "      <td>164</td>\n",
       "      <td>PRJNA997801, PRJNA1094419</td>\n",
       "      <td>Carlino, N., Blanco-Míguez, A., Punčochář, M.,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>MASTER_WP4_TEAGASC_1</td>\n",
       "      <td>156</td>\n",
       "      <td>PRGEB63604, PRJNA1094419</td>\n",
       "      <td>Carlino, N., Blanco-Míguez, A., Punčochář, M.,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>MASTER_WP4_UNINA_2</td>\n",
       "      <td>150</td>\n",
       "      <td>PRJNA897099, PRJNA997801, PRJNA1094419</td>\n",
       "      <td>Carlino, N., Blanco-Míguez, A., Punčochář, M.,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>MortensenS_xxxx</td>\n",
       "      <td>375</td>\n",
       "      <td>PRJNA977472, PRJNA1023503</td>\n",
       "      <td>Carlino, N., Blanco-Míguez, A., Punčochář, M.,...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 Dataset  # Samples  \\\n",
       "0   AlvarezOrdonezA_xxxx        173   \n",
       "5        CrovadoreJ_2017          5   \n",
       "26     MASTER_WP4_CSIC_2        164   \n",
       "31  MASTER_WP4_TEAGASC_1        156   \n",
       "33    MASTER_WP4_UNINA_2        150   \n",
       "36       MortensenS_xxxx        375   \n",
       "\n",
       "                                     Accession number  \\\n",
       "0                            PRJNA997800, PRJNA997801   \n",
       "5   PRJNA358537, PRJNA358558, PRJNA362533, PRJNA36...   \n",
       "26                          PRJNA997801, PRJNA1094419   \n",
       "31                           PRGEB63604, PRJNA1094419   \n",
       "33             PRJNA897099, PRJNA997801, PRJNA1094419   \n",
       "36                          PRJNA977472, PRJNA1023503   \n",
       "\n",
       "                                            Reference  \n",
       "0   Carlino, N., Blanco-Míguez, A., Punčochář, M.,...  \n",
       "5   Crovadore, J., Gérard, F., Chablais, R., Cocha...  \n",
       "26  Carlino, N., Blanco-Míguez, A., Punčochář, M.,...  \n",
       "31  Carlino, N., Blanco-Míguez, A., Punčochář, M.,...  \n",
       "33  Carlino, N., Blanco-Míguez, A., Punčochář, M.,...  \n",
       "36  Carlino, N., Blanco-Míguez, A., Punčochář, M.,...  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Number of rows with multiple accession numbers: 6\n",
      "total number of rows in dataset before splitting: 59\n",
      "number of rows at the end: 69\n"
     ]
    }
   ],
   "source": [
    "# organising carlino_supp\n",
    "    # replacing * with the Carlino paper\n",
    "    # spliting rows with multiple bioproject accession numbers\n",
    "\n",
    "# Replacing '*' in the Reference column with the full citation of the Carlino paper\n",
    "ref = 'Carlino, N., Blanco-Míguez, A., Punčochář, M., Mengoni, C., Pinto, F., Tatti, A., Manghi, P., Armanini, F., Avagliano, M., Barcenilla, C., Breselge, S., Cabrera-Rubio, R., Calvete-Torre, I., Coakley, M., Cobo-Díaz, J. F., De Filippis, F., Dey, H., Leech, J., Klaassens, E. S., Knobloch, S., O’Neil, D., Quijada, N. M., Sabater, C., Skírnisdóttir, S., Valentino, V., Walsh, L., MASTER EU Consortium, Asnicar, F., Fackelmann, G., Heidrich, V., Rota Stabelli, O., Pasolli, E., Arendt, E., Rubino, F., Cotter, P. D., Ruiz, L., Ercolini, D., Morrison, S., Marchart, J., Sessitsch, A., and Zannini, E. (2024). A pan-European study of the human gut microbiome reveals links with diet, health, and lifestyle. Cell, 187(13), 5795–5816.e15. https://doi.org/10.1016/j.cell.2024.05.028'\n",
    "carlino_supp.loc[carlino_supp['Reference'] == '*', 'Reference'] = ref\n",
    "\n",
    "# Splitting rows with multiple bioproject accession numbers\n",
    "# check how many rows have multiple Bioproject accession numbers\n",
    "# Find rows with commas in 'Accession number' and count values after splitting\n",
    "rows_with_commas = carlino_supp[carlino_supp['Accession number'].str.contains(',', na=False)]\n",
    "total_values = sum(len(accession.split(',')) for accession in rows_with_commas['Accession number'])\n",
    "print(f\"Total values after splitting: {total_values}\")\n",
    "\n",
    "print(\"\\nRows with multiple accession numbers:\")\n",
    "display(rows_with_commas)\n",
    "print(f\"\\nNumber of rows with multiple accession numbers: {rows_with_commas.shape[0]}\")\n",
    "print('total number of rows in dataset before splitting:', carlino_supp.shape[0])\n",
    "\n",
    "carlino_supp['Accession number'] = carlino_supp['Accession number'].str.split(',')\n",
    "carlino_supp = carlino_supp.explode('Accession number')\n",
    "# strip white spaces in the 'Accession number' column\n",
    "carlino_supp['Accession number'] = carlino_supp['Accession number'].str.strip()\n",
    "\n",
    "end_rows = carlino_supp.shape[0]\n",
    "print('number of rows at the end:', end_rows)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of rows with References: 2477\n",
      "Total rows in carlino_metadata: 3368\n"
     ]
    }
   ],
   "source": [
    "# Add Reference column from carlino_supp to carlino_metadata\n",
    "# Create a dictionary from carlino_supp, keeping the first occurrence of each project_accession\n",
    "reference_dict = dict(zip(carlino_supp['Accession number'], carlino_supp['Reference']))\n",
    "carlino_metadata['Reference'] = carlino_metadata['project_accession'].map(reference_dict)\n",
    "\n",
    "# Verify the merge by checking if References were added\n",
    "print(\"Number of rows with References:\", carlino_metadata['Reference'].notna().sum())\n",
    "print(\"Total rows in carlino_metadata:\", len(carlino_metadata))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>dataset_name</th>\n",
       "      <th>project_accession</th>\n",
       "      <th>Reference</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>408</th>\n",
       "      <td>KastmanEK_2016</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>799</th>\n",
       "      <td>MASTER_WP4_CSIC_2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2035</th>\n",
       "      <td>PorcellatoD_2016</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2469</th>\n",
       "      <td>WolfeBE_2014</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2493</th>\n",
       "      <td>XieM_2019</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2547</th>\n",
       "      <td>CM_UNINA_FFOOD</td>\n",
       "      <td>PRJNA1206071</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2808</th>\n",
       "      <td>DecadtH_2024</td>\n",
       "      <td>PRJEB64331</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2827</th>\n",
       "      <td>FalardeauJ_2023</td>\n",
       "      <td>PRJNA863305</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2883</th>\n",
       "      <td>FontanaF_2023</td>\n",
       "      <td>PRJNA865096</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3011</th>\n",
       "      <td>FranciosaI_2021</td>\n",
       "      <td>PRJNA636619</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3026</th>\n",
       "      <td>GonzalezOrozcoB_2023</td>\n",
       "      <td>PRJNA843464</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3028</th>\n",
       "      <td>KharnaiorP_2023</td>\n",
       "      <td>PRJNA914873</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3032</th>\n",
       "      <td>LimaC_2020</td>\n",
       "      <td>PRJNA552479</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3046</th>\n",
       "      <td>LopezSanchezR_2023</td>\n",
       "      <td>PRJNA648868</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3050</th>\n",
       "      <td>MagliuloR_2024</td>\n",
       "      <td>PRJNA1084214</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3099</th>\n",
       "      <td>MotaGutierrezJ_2021</td>\n",
       "      <td>PRJNA627182</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3119</th>\n",
       "      <td>OlgaP_2019</td>\n",
       "      <td>PRJNA636820</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3121</th>\n",
       "      <td>OlgaP_2019</td>\n",
       "      <td>PRJNA636837</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3123</th>\n",
       "      <td>OlgaP_2019</td>\n",
       "      <td>PRJNA636891</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3125</th>\n",
       "      <td>OlgaP_2019</td>\n",
       "      <td>PRJNA637016</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3127</th>\n",
       "      <td>OlgaP_2019</td>\n",
       "      <td>PRJNA637018</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3128</th>\n",
       "      <td>QuijadaN_2022</td>\n",
       "      <td>PRJEB48589</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3148</th>\n",
       "      <td>SalgadoTS_2021</td>\n",
       "      <td>PRJNA704713</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3287</th>\n",
       "      <td>ShangpliangH_2023_a</td>\n",
       "      <td>PRJNA914731</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3289</th>\n",
       "      <td>ShangpliangH_2023_b</td>\n",
       "      <td>PRJNA914730</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3291</th>\n",
       "      <td>TomarS_2023</td>\n",
       "      <td>PRJNA763767</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3292</th>\n",
       "      <td>YangC_2021</td>\n",
       "      <td>PRJNA683931</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3318</th>\n",
       "      <td>YasirM_2022</td>\n",
       "      <td>PRJNA816670</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3336</th>\n",
       "      <td>YouL_2022</td>\n",
       "      <td>PRJNA687995</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3359</th>\n",
       "      <td>YuY_2022</td>\n",
       "      <td>PRJNA780248</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              dataset_name project_accession Reference\n",
       "408         KastmanEK_2016               NaN       NaN\n",
       "799      MASTER_WP4_CSIC_2               NaN       NaN\n",
       "2035      PorcellatoD_2016               NaN       NaN\n",
       "2469          WolfeBE_2014               NaN       NaN\n",
       "2493             XieM_2019               NaN       NaN\n",
       "2547        CM_UNINA_FFOOD      PRJNA1206071       NaN\n",
       "2808          DecadtH_2024        PRJEB64331       NaN\n",
       "2827       FalardeauJ_2023       PRJNA863305       NaN\n",
       "2883         FontanaF_2023       PRJNA865096       NaN\n",
       "3011       FranciosaI_2021       PRJNA636619       NaN\n",
       "3026  GonzalezOrozcoB_2023       PRJNA843464       NaN\n",
       "3028       KharnaiorP_2023       PRJNA914873       NaN\n",
       "3032            LimaC_2020       PRJNA552479       NaN\n",
       "3046    LopezSanchezR_2023       PRJNA648868       NaN\n",
       "3050        MagliuloR_2024      PRJNA1084214       NaN\n",
       "3099   MotaGutierrezJ_2021       PRJNA627182       NaN\n",
       "3119            OlgaP_2019       PRJNA636820       NaN\n",
       "3121            OlgaP_2019       PRJNA636837       NaN\n",
       "3123            OlgaP_2019       PRJNA636891       NaN\n",
       "3125            OlgaP_2019       PRJNA637016       NaN\n",
       "3127            OlgaP_2019       PRJNA637018       NaN\n",
       "3128         QuijadaN_2022        PRJEB48589       NaN\n",
       "3148        SalgadoTS_2021       PRJNA704713       NaN\n",
       "3287   ShangpliangH_2023_a       PRJNA914731       NaN\n",
       "3289   ShangpliangH_2023_b       PRJNA914730       NaN\n",
       "3291           TomarS_2023       PRJNA763767       NaN\n",
       "3292            YangC_2021       PRJNA683931       NaN\n",
       "3318           YasirM_2022       PRJNA816670       NaN\n",
       "3336             YouL_2022       PRJNA687995       NaN\n",
       "3359              YuY_2022       PRJNA780248       NaN"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# checking the missing references\n",
    "missing_refs = carlino_metadata[\n",
    "    (carlino_metadata.Reference.isna()) & \n",
    "    (carlino_metadata['fermented/non-fermented'] == 'F')\n",
    "][['dataset_name', 'project_accession','Reference']].drop_duplicates()\n",
    "\n",
    "# check if the dataset_name is in the Fermentos metadata\n",
    "missing_refs = missing_refs[missing_refs['dataset_name'].apply(lambda x: dataset['mag_id'].str.contains(x).any())]\n",
    "display(missing_refs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "# saving the missing references table to a file\n",
    "# will locate references manually\n",
    "missing_refs.to_csv('data/missing_references.tsv', sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>dataset_name</th>\n",
       "      <th>project_accession</th>\n",
       "      <th>Reference</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>KastmanEK_2016</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Kastman, E.K., Kamelamela, N., Norville, J.W.,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>PorcellatoD_2016</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Porcellato, D., and Skeie, S.B. (2016). Bacter...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>WolfeBE_2014</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Wolfe, B.E., Button, J.E., Santarelli, M., and...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>XieM_2019</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Xie, M., Wu, J., An, F., Yue, X., Tao, D., Wu,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CM_UNINA_FFOOD</td>\n",
       "      <td>PRJNA1206071</td>\n",
       "      <td>Carlino, N., Blanco-Míguez, A., Punčochář, M.,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>DecadtH_2024</td>\n",
       "      <td>PRJEB64331</td>\n",
       "      <td>Decadt H, Vermote L, Díaz-Muñoz C, Weckx S, De...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>FalardeauJ_2023</td>\n",
       "      <td>PRJNA863305</td>\n",
       "      <td>Falardeau J, Yildiz E, Yan Y, Castellarin SD, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>FontanaF_2023</td>\n",
       "      <td>PRJNA865096</td>\n",
       "      <td>Fontana F, Longhi G, Alessandri G, Lugli GA, M...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>FranciosaI_2021</td>\n",
       "      <td>PRJNA636619</td>\n",
       "      <td>Franciosa, I., Ferrocino, I., Giordano, M., Mo...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>GonzalezOrozcoB_2023</td>\n",
       "      <td>PRJNA843464</td>\n",
       "      <td>González-Orozco BD, García-Cano I, Escobar-Zep...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>KharnaiorP_2023</td>\n",
       "      <td>PRJNA914873</td>\n",
       "      <td>Kharnaior P, Tamang JP. Microbiome and metabol...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>LimaC_2020</td>\n",
       "      <td>PRJNA552479</td>\n",
       "      <td>de C Lima CO, Vaz ABM, De Castro GM, Lobo F, S...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>LopezSanchezR_2023</td>\n",
       "      <td>PRJNA648868</td>\n",
       "      <td>López-Sánchez R, Hernández-Oaxaca D, Escobar-Z...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>MagliuloR_2024</td>\n",
       "      <td>PRJNA1084214</td>\n",
       "      <td>Magarelli M, Novielli P, De Filippis F, Magliu...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>MotaGutierrezJ_2021</td>\n",
       "      <td>PRJNA627182</td>\n",
       "      <td>Mota-Gutierrez J, Ferrocino I, Giordano M, Sua...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>OlgaP_2019</td>\n",
       "      <td>PRJNA636820</td>\n",
       "      <td>Lee I, Barh D, Podolich O, Brenig B, Tiwari S,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>OlgaP_2019</td>\n",
       "      <td>PRJNA636837</td>\n",
       "      <td>Lee I, Barh D, Podolich O, Brenig B, Tiwari S,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>OlgaP_2019</td>\n",
       "      <td>PRJNA636891</td>\n",
       "      <td>Lee I, Barh D, Podolich O, Brenig B, Tiwari S,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>OlgaP_2019</td>\n",
       "      <td>PRJNA637016</td>\n",
       "      <td>Lee I, Barh D, Podolich O, Brenig B, Tiwari S,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>OlgaP_2019</td>\n",
       "      <td>PRJNA637018</td>\n",
       "      <td>Lee I, Barh D, Podolich O, Brenig B, Tiwari S,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>QuijadaN_2022</td>\n",
       "      <td>PRJEB48589</td>\n",
       "      <td>Quijada NM, Dzieciol M, Schmitz-Esser S, Wagne...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>SalgadoTS_2021</td>\n",
       "      <td>PRJNA704713</td>\n",
       "      <td>Tenorio-Salgado S, Castelán-Sánchez HG, Dávila...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>ShangpliangH_2023_a</td>\n",
       "      <td>PRJNA914731</td>\n",
       "      <td>Shangpliang HNJ, Tamang JP. Metagenome-assembl...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>ShangpliangH_2023_b</td>\n",
       "      <td>PRJNA914730</td>\n",
       "      <td>Shangpliang HNJ, Tamang JP. Metagenome-assembl...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>TomarS_2023</td>\n",
       "      <td>PRJNA763767</td>\n",
       "      <td>Tomar S, Mitra D, Kumar G, Kashyap P, Sharma M...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>YangC_2021</td>\n",
       "      <td>PRJNA683931</td>\n",
       "      <td>Yang, C., You, L., Kwok, L.-Y., Jin, H., Peng,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>YasirM_2022</td>\n",
       "      <td>PRJNA816670</td>\n",
       "      <td>Muhammad Yasir, Ibrahim A. Al-Zahrani, Fehmida...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>YouL_2022</td>\n",
       "      <td>PRJNA687995</td>\n",
       "      <td>You, L., Yang, C., Jin, H., Kwok, L.-Y., Sun, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>YuY_2022</td>\n",
       "      <td>PRJNA780248</td>\n",
       "      <td>Yu Y, Li L, Xu Y, Li H, Yu Y, Xu Z. Metagenomi...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            dataset_name project_accession  \\\n",
       "0         KastmanEK_2016               NaN   \n",
       "1       PorcellatoD_2016               NaN   \n",
       "2           WolfeBE_2014               NaN   \n",
       "3              XieM_2019               NaN   \n",
       "4         CM_UNINA_FFOOD      PRJNA1206071   \n",
       "5           DecadtH_2024        PRJEB64331   \n",
       "6        FalardeauJ_2023       PRJNA863305   \n",
       "7          FontanaF_2023       PRJNA865096   \n",
       "8        FranciosaI_2021       PRJNA636619   \n",
       "9   GonzalezOrozcoB_2023       PRJNA843464   \n",
       "10       KharnaiorP_2023       PRJNA914873   \n",
       "11            LimaC_2020       PRJNA552479   \n",
       "12    LopezSanchezR_2023       PRJNA648868   \n",
       "13        MagliuloR_2024      PRJNA1084214   \n",
       "14   MotaGutierrezJ_2021       PRJNA627182   \n",
       "15            OlgaP_2019       PRJNA636820   \n",
       "16            OlgaP_2019       PRJNA636837   \n",
       "17            OlgaP_2019       PRJNA636891   \n",
       "18            OlgaP_2019       PRJNA637016   \n",
       "19            OlgaP_2019       PRJNA637018   \n",
       "20         QuijadaN_2022        PRJEB48589   \n",
       "21        SalgadoTS_2021       PRJNA704713   \n",
       "22   ShangpliangH_2023_a       PRJNA914731   \n",
       "23   ShangpliangH_2023_b       PRJNA914730   \n",
       "24           TomarS_2023       PRJNA763767   \n",
       "25            YangC_2021       PRJNA683931   \n",
       "26           YasirM_2022       PRJNA816670   \n",
       "27             YouL_2022       PRJNA687995   \n",
       "28              YuY_2022       PRJNA780248   \n",
       "\n",
       "                                            Reference  \n",
       "0   Kastman, E.K., Kamelamela, N., Norville, J.W.,...  \n",
       "1   Porcellato, D., and Skeie, S.B. (2016). Bacter...  \n",
       "2   Wolfe, B.E., Button, J.E., Santarelli, M., and...  \n",
       "3   Xie, M., Wu, J., An, F., Yue, X., Tao, D., Wu,...  \n",
       "4   Carlino, N., Blanco-Míguez, A., Punčochář, M.,...  \n",
       "5   Decadt H, Vermote L, Díaz-Muñoz C, Weckx S, De...  \n",
       "6   Falardeau J, Yildiz E, Yan Y, Castellarin SD, ...  \n",
       "7   Fontana F, Longhi G, Alessandri G, Lugli GA, M...  \n",
       "8   Franciosa, I., Ferrocino, I., Giordano, M., Mo...  \n",
       "9   González-Orozco BD, García-Cano I, Escobar-Zep...  \n",
       "10  Kharnaior P, Tamang JP. Microbiome and metabol...  \n",
       "11  de C Lima CO, Vaz ABM, De Castro GM, Lobo F, S...  \n",
       "12  López-Sánchez R, Hernández-Oaxaca D, Escobar-Z...  \n",
       "13  Magarelli M, Novielli P, De Filippis F, Magliu...  \n",
       "14  Mota-Gutierrez J, Ferrocino I, Giordano M, Sua...  \n",
       "15  Lee I, Barh D, Podolich O, Brenig B, Tiwari S,...  \n",
       "16  Lee I, Barh D, Podolich O, Brenig B, Tiwari S,...  \n",
       "17  Lee I, Barh D, Podolich O, Brenig B, Tiwari S,...  \n",
       "18  Lee I, Barh D, Podolich O, Brenig B, Tiwari S,...  \n",
       "19  Lee I, Barh D, Podolich O, Brenig B, Tiwari S,...  \n",
       "20  Quijada NM, Dzieciol M, Schmitz-Esser S, Wagne...  \n",
       "21  Tenorio-Salgado S, Castelán-Sánchez HG, Dávila...  \n",
       "22  Shangpliang HNJ, Tamang JP. Metagenome-assembl...  \n",
       "23  Shangpliang HNJ, Tamang JP. Metagenome-assembl...  \n",
       "24  Tomar S, Mitra D, Kumar G, Kashyap P, Sharma M...  \n",
       "25  Yang, C., You, L., Kwok, L.-Y., Jin, H., Peng,...  \n",
       "26  Muhammad Yasir, Ibrahim A. Al-Zahrani, Fehmida...  \n",
       "27  You, L., Yang, C., Jin, H., Kwok, L.-Y., Sun, ...  \n",
       "28  Yu Y, Li L, Xu Y, Li H, Yu Y, Xu Z. Metagenomi...  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Loading back the missing_ref table after manual curation. \n",
    "# all samples now have references but not all have a bioproject since they are not all from NCBI. \n",
    "missing_refs = pd.read_csv('data/missing_references.tsv', sep='\\t')\n",
    "display(missing_refs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Add the curated references to the carlino_metadata\n",
    "missing_refs_dict = dict(zip(missing_refs['dataset_name'], missing_refs['Reference']))\n",
    "carlino_metadata['Reference'] = carlino_metadata['dataset_name'].map(missing_refs_dict)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>dataset_name</th>\n",
       "      <th>project_accession</th>\n",
       "      <th>Reference</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AlvarezOrdonezA_xxxx</td>\n",
       "      <td>PRJNA997800</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>72</th>\n",
       "      <td>AlvarezOrdonezA_xxxx</td>\n",
       "      <td>PRJNA997801</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>173</th>\n",
       "      <td>ArikanM_2020</td>\n",
       "      <td>PRJEB22200</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>185</th>\n",
       "      <td>BertuzziAS_2018</td>\n",
       "      <td>PRJEB15423</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>227</th>\n",
       "      <td>ChaconVargasK_2020</td>\n",
       "      <td>PRJNA603591</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>232</th>\n",
       "      <td>CM_INJERA</td>\n",
       "      <td>PRJNA504891</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>239</th>\n",
       "      <td>DeFilippisF_xxxx</td>\n",
       "      <td>PRJNA1094419</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>358</th>\n",
       "      <td>DeRoosJ_2020</td>\n",
       "      <td>PRJEB28363</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>364</th>\n",
       "      <td>DuR_2020</td>\n",
       "      <td>PRJDB9193</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>367</th>\n",
       "      <td>DuruIC_2018</td>\n",
       "      <td>PRJEB23938</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>373</th>\n",
       "      <td>EinsonEJ_2018</td>\n",
       "      <td>PRJNA336518</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>378</th>\n",
       "      <td>EscobarZepedaA_2016</td>\n",
       "      <td>PRJNA286900</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>379</th>\n",
       "      <td>FerrocinoI_2018</td>\n",
       "      <td>PRJNA352236</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>395</th>\n",
       "      <td>HeilCS_2018</td>\n",
       "      <td>PRJNA390460</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>428</th>\n",
       "      <td>KumarJ_2019</td>\n",
       "      <td>PRJNA529966</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>432</th>\n",
       "      <td>LandisEA_2021</td>\n",
       "      <td>PRJNA589612</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>472</th>\n",
       "      <td>LeechJ_2020</td>\n",
       "      <td>PRJEB35321</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>530</th>\n",
       "      <td>LeechJ_xxxx</td>\n",
       "      <td>PRJNA1052643</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>633</th>\n",
       "      <td>LiZ_2018</td>\n",
       "      <td>PRJNA433075</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>639</th>\n",
       "      <td>LiZ_2019</td>\n",
       "      <td>PRJNA473190</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>647</th>\n",
       "      <td>LordanR_2019</td>\n",
       "      <td>PRJEB30083</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>652</th>\n",
       "      <td>MASTER_WP4_CSIC_1</td>\n",
       "      <td>PRJNA997801</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>766</th>\n",
       "      <td>MASTER_WP4_CSIC_2</td>\n",
       "      <td>PRJNA997801</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>799</th>\n",
       "      <td>MASTER_WP4_CSIC_2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>808</th>\n",
       "      <td>MASTER_WP4_CSIC_2</td>\n",
       "      <td>PRJNA1094419</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>930</th>\n",
       "      <td>MASTER_WP4_FFoQSI_1</td>\n",
       "      <td>PRJNA997801</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1002</th>\n",
       "      <td>MASTER_WP4_FFoQSI_2</td>\n",
       "      <td>PRJNA997801</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1103</th>\n",
       "      <td>MASTER_WP4_FFoQSI_3</td>\n",
       "      <td>PRJNA997801</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1178</th>\n",
       "      <td>MASTER_WP4_TEAGASC_1</td>\n",
       "      <td>PRJNA1094419</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1180</th>\n",
       "      <td>MASTER_WP4_TEAGASC_1</td>\n",
       "      <td>PRGEB63604</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1333</th>\n",
       "      <td>MASTER_WP4_UNINA_1</td>\n",
       "      <td>PRJNA1094419</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1429</th>\n",
       "      <td>MASTER_WP4_UNINA_2</td>\n",
       "      <td>PRJNA997801</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1594</th>\n",
       "      <td>MilaniC_2019</td>\n",
       "      <td>PRJNA482503</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1598</th>\n",
       "      <td>MortensenS_xxxx</td>\n",
       "      <td>PRJNA977472</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1603</th>\n",
       "      <td>MortensenS_xxxx</td>\n",
       "      <td>PRJNA1023503</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1975</th>\n",
       "      <td>PasolliE_2020</td>\n",
       "      <td>PRJNA603575</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1999</th>\n",
       "      <td>PfeferT_xxxx</td>\n",
       "      <td>PRJNA430402</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2047</th>\n",
       "      <td>PothakosV_2020</td>\n",
       "      <td>PRJEB24129</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2053</th>\n",
       "      <td>QuigleyL_2016</td>\n",
       "      <td>PRJEB6952</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2066</th>\n",
       "      <td>SomervilleV_2019</td>\n",
       "      <td>PRJNA454439</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2067</th>\n",
       "      <td>SternesPR_2017</td>\n",
       "      <td>PRJNA305659</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2087</th>\n",
       "      <td>SulaimanJ_2014</td>\n",
       "      <td>PRJNA185981</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2094</th>\n",
       "      <td>VerceM_2019</td>\n",
       "      <td>PRJEB21603</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2098</th>\n",
       "      <td>WalshAM_2016</td>\n",
       "      <td>PRJEB15432</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2116</th>\n",
       "      <td>WalshAM_2017</td>\n",
       "      <td>PRJEB20873</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2126</th>\n",
       "      <td>WalshAM_2020</td>\n",
       "      <td>PRJEB32768</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2203</th>\n",
       "      <td>WalshL_xxxx</td>\n",
       "      <td>PRJEB65292</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2494</th>\n",
       "      <td>YaoG_2017</td>\n",
       "      <td>PRJNA340315</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2524</th>\n",
       "      <td>YasirM_2020</td>\n",
       "      <td>PRJEB36988</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2526</th>\n",
       "      <td>YulandiA_2020</td>\n",
       "      <td>PRJNA605305</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2528</th>\n",
       "      <td>ZhaoCC_2020</td>\n",
       "      <td>PRJNA594855</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              dataset_name project_accession Reference\n",
       "2     AlvarezOrdonezA_xxxx       PRJNA997800       NaN\n",
       "72    AlvarezOrdonezA_xxxx       PRJNA997801       NaN\n",
       "173           ArikanM_2020        PRJEB22200       NaN\n",
       "185        BertuzziAS_2018        PRJEB15423       NaN\n",
       "227     ChaconVargasK_2020       PRJNA603591       NaN\n",
       "232              CM_INJERA       PRJNA504891       NaN\n",
       "239       DeFilippisF_xxxx      PRJNA1094419       NaN\n",
       "358           DeRoosJ_2020        PRJEB28363       NaN\n",
       "364               DuR_2020         PRJDB9193       NaN\n",
       "367            DuruIC_2018        PRJEB23938       NaN\n",
       "373          EinsonEJ_2018       PRJNA336518       NaN\n",
       "378    EscobarZepedaA_2016       PRJNA286900       NaN\n",
       "379        FerrocinoI_2018       PRJNA352236       NaN\n",
       "395            HeilCS_2018       PRJNA390460       NaN\n",
       "428            KumarJ_2019       PRJNA529966       NaN\n",
       "432          LandisEA_2021       PRJNA589612       NaN\n",
       "472            LeechJ_2020        PRJEB35321       NaN\n",
       "530            LeechJ_xxxx      PRJNA1052643       NaN\n",
       "633               LiZ_2018       PRJNA433075       NaN\n",
       "639               LiZ_2019       PRJNA473190       NaN\n",
       "647           LordanR_2019        PRJEB30083       NaN\n",
       "652      MASTER_WP4_CSIC_1       PRJNA997801       NaN\n",
       "766      MASTER_WP4_CSIC_2       PRJNA997801       NaN\n",
       "799      MASTER_WP4_CSIC_2               NaN       NaN\n",
       "808      MASTER_WP4_CSIC_2      PRJNA1094419       NaN\n",
       "930    MASTER_WP4_FFoQSI_1       PRJNA997801       NaN\n",
       "1002   MASTER_WP4_FFoQSI_2       PRJNA997801       NaN\n",
       "1103   MASTER_WP4_FFoQSI_3       PRJNA997801       NaN\n",
       "1178  MASTER_WP4_TEAGASC_1      PRJNA1094419       NaN\n",
       "1180  MASTER_WP4_TEAGASC_1        PRGEB63604       NaN\n",
       "1333    MASTER_WP4_UNINA_1      PRJNA1094419       NaN\n",
       "1429    MASTER_WP4_UNINA_2       PRJNA997801       NaN\n",
       "1594          MilaniC_2019       PRJNA482503       NaN\n",
       "1598       MortensenS_xxxx       PRJNA977472       NaN\n",
       "1603       MortensenS_xxxx      PRJNA1023503       NaN\n",
       "1975         PasolliE_2020       PRJNA603575       NaN\n",
       "1999          PfeferT_xxxx       PRJNA430402       NaN\n",
       "2047        PothakosV_2020        PRJEB24129       NaN\n",
       "2053         QuigleyL_2016         PRJEB6952       NaN\n",
       "2066      SomervilleV_2019       PRJNA454439       NaN\n",
       "2067        SternesPR_2017       PRJNA305659       NaN\n",
       "2087        SulaimanJ_2014       PRJNA185981       NaN\n",
       "2094           VerceM_2019        PRJEB21603       NaN\n",
       "2098          WalshAM_2016        PRJEB15432       NaN\n",
       "2116          WalshAM_2017        PRJEB20873       NaN\n",
       "2126          WalshAM_2020        PRJEB32768       NaN\n",
       "2203           WalshL_xxxx        PRJEB65292       NaN\n",
       "2494             YaoG_2017       PRJNA340315       NaN\n",
       "2524           YasirM_2020        PRJEB36988       NaN\n",
       "2526         YulandiA_2020       PRJNA605305       NaN\n",
       "2528           ZhaoCC_2020       PRJNA594855       NaN"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# checking the missing references again\n",
    "missing_refs = carlino_metadata[\n",
    "    (carlino_metadata.Reference.isna()) & \n",
    "    (carlino_metadata['fermented/non-fermented'] == 'F')\n",
    "][['dataset_name', 'project_accession','Reference']].drop_duplicates()\n",
    "\n",
    "# check if the dataset_name is in the Fermentos metadata\n",
    "missing_refs = missing_refs[missing_refs['dataset_name'].apply(lambda x: dataset['mag_id'].str.contains(x).any())]\n",
    "display(missing_refs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of rows with References: 708\n",
      "Total rows in carlino_metadata: 3368\n"
     ]
    }
   ],
   "source": [
    "print(\"Number of rows with References:\", carlino_metadata['Reference'].notna().sum())\n",
    "print(\"Total rows in carlino_metadata:\", len(carlino_metadata))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "2       True\n",
       "72      True\n",
       "173     True\n",
       "185     True\n",
       "227     True\n",
       "232     True\n",
       "239     True\n",
       "358     True\n",
       "364     True\n",
       "367     True\n",
       "373     True\n",
       "378     True\n",
       "379     True\n",
       "395     True\n",
       "428     True\n",
       "432     True\n",
       "472     True\n",
       "530     True\n",
       "633     True\n",
       "639     True\n",
       "647     True\n",
       "652     True\n",
       "766     True\n",
       "799     True\n",
       "808     True\n",
       "930     True\n",
       "1002    True\n",
       "1103    True\n",
       "1178    True\n",
       "1180    True\n",
       "1333    True\n",
       "1429    True\n",
       "1594    True\n",
       "1598    True\n",
       "1603    True\n",
       "1975    True\n",
       "1999    True\n",
       "2047    True\n",
       "2053    True\n",
       "2066    True\n",
       "2067    True\n",
       "2087    True\n",
       "2094    True\n",
       "2098    True\n",
       "2116    True\n",
       "2126    True\n",
       "2203    True\n",
       "2494    True\n",
       "2524    True\n",
       "2526    True\n",
       "2528    True\n",
       "Name: dataset_name, dtype: bool"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# compare the values in dataset_name from missing_refs to Dataset in carlino_supp. I want to finf if any of them are in there\n",
    "missing_refs['dataset_name'].apply(lambda x: carlino_supp['Dataset'].str.contains(x).any()) "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# issues with references \n",
    "It looks as though mapping the references shifts the data around. I will try and solve this in a different way. <br>\n",
    "I need to reload the supp and metadata files. <br>\n",
    "I will add the missing refs to the supp table first.<br>\n",
    "I will mapp based on three columns (dataset_name, bioproject and referecnce)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>dataset_name</th>\n",
       "      <th>sample_id</th>\n",
       "      <th>macrocategory</th>\n",
       "      <th>category</th>\n",
       "      <th>type</th>\n",
       "      <th>subtype</th>\n",
       "      <th>commercial_name</th>\n",
       "      <th>fermented/non-fermented</th>\n",
       "      <th>country</th>\n",
       "      <th>sample_accession</th>\n",
       "      <th>...</th>\n",
       "      <th>median_read_len</th>\n",
       "      <th>mean_read_len</th>\n",
       "      <th>max_read_len</th>\n",
       "      <th>n_contigs</th>\n",
       "      <th>n_MAGs_MQ_prok</th>\n",
       "      <th>n_MAGs_HQ_prok</th>\n",
       "      <th>n_MAGs_MQ_euk</th>\n",
       "      <th>n_MAGs_HQ_euk</th>\n",
       "      <th>filtered</th>\n",
       "      <th>curator</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AlvarezOrdonezA_xxxx</td>\n",
       "      <td>C01-16</td>\n",
       "      <td>food</td>\n",
       "      <td>meat</td>\n",
       "      <td>cured_meat_before_ripening</td>\n",
       "      <td>cured_meat_before_ripening</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NF</td>\n",
       "      <td>ESP</td>\n",
       "      <td>SAMN36689796</td>\n",
       "      <td>...</td>\n",
       "      <td>151.0</td>\n",
       "      <td>146.200505</td>\n",
       "      <td>151.0</td>\n",
       "      <td>62700</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>no</td>\n",
       "      <td>Niccolò_Carlino</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AlvarezOrdonezA_xxxx</td>\n",
       "      <td>C01-17</td>\n",
       "      <td>food</td>\n",
       "      <td>meat</td>\n",
       "      <td>cured_meat_before_ripening</td>\n",
       "      <td>cured_meat_before_ripening</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NF</td>\n",
       "      <td>ESP</td>\n",
       "      <td>SAMN36689797</td>\n",
       "      <td>...</td>\n",
       "      <td>151.0</td>\n",
       "      <td>147.266426</td>\n",
       "      <td>151.0</td>\n",
       "      <td>98756</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>no</td>\n",
       "      <td>Niccolò_Carlino</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AlvarezOrdonezA_xxxx</td>\n",
       "      <td>C01-18</td>\n",
       "      <td>food</td>\n",
       "      <td>fermented_meat</td>\n",
       "      <td>meat</td>\n",
       "      <td>cured_meat_after_ripening</td>\n",
       "      <td>NaN</td>\n",
       "      <td>F</td>\n",
       "      <td>ESP</td>\n",
       "      <td>SAMN36689824</td>\n",
       "      <td>...</td>\n",
       "      <td>151.0</td>\n",
       "      <td>147.760911</td>\n",
       "      <td>151.0</td>\n",
       "      <td>103911</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>no</td>\n",
       "      <td>Niccolò_Carlino</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AlvarezOrdonezA_xxxx</td>\n",
       "      <td>C01-19</td>\n",
       "      <td>food</td>\n",
       "      <td>fermented_meat</td>\n",
       "      <td>meat</td>\n",
       "      <td>cured_meat_after_ripening</td>\n",
       "      <td>NaN</td>\n",
       "      <td>F</td>\n",
       "      <td>ESP</td>\n",
       "      <td>SAMN36689825</td>\n",
       "      <td>...</td>\n",
       "      <td>151.0</td>\n",
       "      <td>149.200530</td>\n",
       "      <td>151.0</td>\n",
       "      <td>134121</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>no</td>\n",
       "      <td>Niccolò_Carlino</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AlvarezOrdonezA_xxxx</td>\n",
       "      <td>C02-16</td>\n",
       "      <td>food</td>\n",
       "      <td>meat</td>\n",
       "      <td>cured_meat_before_ripening</td>\n",
       "      <td>cured_meat_before_ripening</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NF</td>\n",
       "      <td>ESP</td>\n",
       "      <td>SAMN36689798</td>\n",
       "      <td>...</td>\n",
       "      <td>151.0</td>\n",
       "      <td>148.854534</td>\n",
       "      <td>151.0</td>\n",
       "      <td>37597</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>no</td>\n",
       "      <td>Niccolò_Carlino</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 32 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           dataset_name sample_id macrocategory        category  \\\n",
       "0  AlvarezOrdonezA_xxxx    C01-16          food            meat   \n",
       "1  AlvarezOrdonezA_xxxx    C01-17          food            meat   \n",
       "2  AlvarezOrdonezA_xxxx    C01-18          food  fermented_meat   \n",
       "3  AlvarezOrdonezA_xxxx    C01-19          food  fermented_meat   \n",
       "4  AlvarezOrdonezA_xxxx    C02-16          food            meat   \n",
       "\n",
       "                         type                     subtype commercial_name  \\\n",
       "0  cured_meat_before_ripening  cured_meat_before_ripening             NaN   \n",
       "1  cured_meat_before_ripening  cured_meat_before_ripening             NaN   \n",
       "2                        meat   cured_meat_after_ripening             NaN   \n",
       "3                        meat   cured_meat_after_ripening             NaN   \n",
       "4  cured_meat_before_ripening  cured_meat_before_ripening             NaN   \n",
       "\n",
       "  fermented/non-fermented country sample_accession  ... median_read_len  \\\n",
       "0                      NF     ESP     SAMN36689796  ...           151.0   \n",
       "1                      NF     ESP     SAMN36689797  ...           151.0   \n",
       "2                       F     ESP     SAMN36689824  ...           151.0   \n",
       "3                       F     ESP     SAMN36689825  ...           151.0   \n",
       "4                      NF     ESP     SAMN36689798  ...           151.0   \n",
       "\n",
       "  mean_read_len max_read_len n_contigs n_MAGs_MQ_prok n_MAGs_HQ_prok  \\\n",
       "0    146.200505        151.0     62700              0              0   \n",
       "1    147.266426        151.0     98756              0              0   \n",
       "2    147.760911        151.0    103911              0              0   \n",
       "3    149.200530        151.0    134121              0              0   \n",
       "4    148.854534        151.0     37597              1              1   \n",
       "\n",
       "  n_MAGs_MQ_euk n_MAGs_HQ_euk filtered          curator  \n",
       "0             0             0       no  Niccolò_Carlino  \n",
       "1             0             0       no  Niccolò_Carlino  \n",
       "2             0             0       no  Niccolò_Carlino  \n",
       "3             0             0       no  Niccolò_Carlino  \n",
       "4             0             0       no  Niccolò_Carlino  \n",
       "\n",
       "[5 rows x 32 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Dataset</th>\n",
       "      <th># Samples</th>\n",
       "      <th>Accession number</th>\n",
       "      <th>Reference</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AlvarezOrdonezA_xxxx</td>\n",
       "      <td>173</td>\n",
       "      <td>PRJNA997800, PRJNA997801</td>\n",
       "      <td>*</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ArikanM_2020</td>\n",
       "      <td>12</td>\n",
       "      <td>PRJEB22200</td>\n",
       "      <td>Arıkan, M., Mitchell, A.L., Finn, R.D., and Gü...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>BertuzziAS_2018</td>\n",
       "      <td>42</td>\n",
       "      <td>PRJEB15423</td>\n",
       "      <td>Bertuzzi, A.S., Walsh, A.M., Sheehan, J.J., Co...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ChaconVargasK_2020</td>\n",
       "      <td>5</td>\n",
       "      <td>PRJNA603591</td>\n",
       "      <td>Chacón-Vargas, K., Torres, J., Giles-Gómez, M....</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CM_INJERA</td>\n",
       "      <td>2</td>\n",
       "      <td>PRJNA504891</td>\n",
       "      <td>Manara, S., Selma-Royo, M., Huang, K.D., Asnic...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                Dataset  # Samples          Accession number  \\\n",
       "0  AlvarezOrdonezA_xxxx        173  PRJNA997800, PRJNA997801   \n",
       "1          ArikanM_2020         12                PRJEB22200   \n",
       "2       BertuzziAS_2018         42                PRJEB15423   \n",
       "3    ChaconVargasK_2020          5               PRJNA603591   \n",
       "4             CM_INJERA          2               PRJNA504891   \n",
       "\n",
       "                                           Reference  \n",
       "0                                                  *  \n",
       "1  Arıkan, M., Mitchell, A.L., Finn, R.D., and Gü...  \n",
       "2  Bertuzzi, A.S., Walsh, A.M., Sheehan, J.J., Co...  \n",
       "3  Chacón-Vargas, K., Torres, J., Giles-Gómez, M....  \n",
       "4  Manara, S., Selma-Royo, M., Huang, K.D., Asnic...  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>dataset_name</th>\n",
       "      <th>project_accession</th>\n",
       "      <th>Reference</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>KastmanEK_2016</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Kastman, E.K., Kamelamela, N., Norville, J.W.,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>PorcellatoD_2016</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Porcellato, D., and Skeie, S.B. (2016). Bacter...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>WolfeBE_2014</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Wolfe, B.E., Button, J.E., Santarelli, M., and...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>XieM_2019</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Xie, M., Wu, J., An, F., Yue, X., Tao, D., Wu,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CM_UNINA_FFOOD</td>\n",
       "      <td>PRJNA1206071</td>\n",
       "      <td>Carlino, N., Blanco-Míguez, A., Punčochář, M.,...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       dataset_name project_accession  \\\n",
       "0    KastmanEK_2016               NaN   \n",
       "1  PorcellatoD_2016               NaN   \n",
       "2      WolfeBE_2014               NaN   \n",
       "3         XieM_2019               NaN   \n",
       "4    CM_UNINA_FFOOD      PRJNA1206071   \n",
       "\n",
       "                                           Reference  \n",
       "0  Kastman, E.K., Kamelamela, N., Norville, J.W.,...  \n",
       "1  Porcellato, D., and Skeie, S.B. (2016). Bacter...  \n",
       "2  Wolfe, B.E., Button, J.E., Santarelli, M., and...  \n",
       "3  Xie, M., Wu, J., An, F., Yue, X., Tao, D., Wu,...  \n",
       "4  Carlino, N., Blanco-Míguez, A., Punčochář, M.,...  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "carlino_metadata = pd.read_csv('data/cFMD_metadata.tsv', sep='\\t')\n",
    "carlino_supp = pd.read_excel('data/Carlino_supp.xlsx') #requires openpyxl to be installed\n",
    "missing_refs = pd.read_csv('data/missing_references.tsv', sep='\\t')\n",
    "\n",
    "display(carlino_metadata.head())\n",
    "display(carlino_supp.head())    \n",
    "display(missing_refs.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "np.int64(4)"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>dataset_name</th>\n",
       "      <th>project_accession</th>\n",
       "      <th>Reference</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>KastmanEK_2016</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Kastman, E.K., Kamelamela, N., Norville, J.W.,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>PorcellatoD_2016</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Porcellato, D., and Skeie, S.B. (2016). Bacter...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>WolfeBE_2014</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Wolfe, B.E., Button, J.E., Santarelli, M., and...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>XieM_2019</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Xie, M., Wu, J., An, F., Yue, X., Tao, D., Wu,...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       dataset_name project_accession  \\\n",
       "0    KastmanEK_2016               NaN   \n",
       "1  PorcellatoD_2016               NaN   \n",
       "2      WolfeBE_2014               NaN   \n",
       "3         XieM_2019               NaN   \n",
       "\n",
       "                                           Reference  \n",
       "0  Kastman, E.K., Kamelamela, N., Norville, J.W.,...  \n",
       "1  Porcellato, D., and Skeie, S.B. (2016). Bacter...  \n",
       "2  Wolfe, B.E., Button, J.E., Santarelli, M., and...  \n",
       "3  Xie, M., Wu, J., An, F., Yue, X., Tao, D., Wu,...  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Dataset</th>\n",
       "      <th># Samples</th>\n",
       "      <th>Accession number</th>\n",
       "      <th>Reference</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>KastmanEK_2016</td>\n",
       "      <td>18</td>\n",
       "      <td>mgp16074</td>\n",
       "      <td>Kastman, E.K., Kamelamela, N., Norville, J.W.,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>PorcellatoD_2016</td>\n",
       "      <td>12</td>\n",
       "      <td>mgp14606</td>\n",
       "      <td>Porcellato, D., and Skeie, S.B. (2016). Bacter...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53</th>\n",
       "      <td>WolfeBE_2014</td>\n",
       "      <td>24</td>\n",
       "      <td>mgp3362</td>\n",
       "      <td>Wolfe, B.E., Button, J.E., Santarelli, M., and...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54</th>\n",
       "      <td>XieM_2019</td>\n",
       "      <td>1</td>\n",
       "      <td>CRA001049</td>\n",
       "      <td>Xie, M., Wu, J., An, F., Yue, X., Tao, D., Wu,...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             Dataset  # Samples Accession number  \\\n",
       "15    KastmanEK_2016         18         mgp16074   \n",
       "40  PorcellatoD_2016         12         mgp14606   \n",
       "53      WolfeBE_2014         24          mgp3362   \n",
       "54         XieM_2019          1        CRA001049   \n",
       "\n",
       "                                            Reference  \n",
       "15  Kastman, E.K., Kamelamela, N., Norville, J.W.,...  \n",
       "40  Porcellato, D., and Skeie, S.B. (2016). Bacter...  \n",
       "53  Wolfe, B.E., Button, J.E., Santarelli, M., and...  \n",
       "54  Xie, M., Wu, J., An, F., Yue, X., Tao, D., Wu,...  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# checking for cross matches bewteen the missing references and the carlino_supp\n",
    "display(missing_refs['dataset_name'].apply(lambda x: carlino_supp['Dataset'].str.contains(x).any()).sum() )\n",
    "\n",
    "# display the rows that have a match in both datasets\n",
    "display(missing_refs[missing_refs['dataset_name'].apply(lambda x: carlino_supp['Dataset'].str.contains(x).any())])\n",
    "display(carlino_supp[carlino_supp['Dataset'].apply(lambda x: missing_refs['dataset_name'].str.contains(x).any())])\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The Accession number in carlino_spp does not match the correct column in carlino_metadata. <br>\n",
    "To ensure proper mapping, I will remove the rows from carlino_supp before concatenating. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of rows in the concatenated dataframe: 84\n",
      "Sum of the two dataframes: 84\n"
     ]
    }
   ],
   "source": [
    "# remove # Samples\tcolumns from carlino_supp\n",
    "carlino_supp_updated = carlino_supp.drop('# Samples', axis=1)\n",
    "\n",
    "# remove rows with dataset_name that are in missing_refs\n",
    "carlino_supp_updated = carlino_supp_updated[~carlino_supp_updated['Dataset'].isin(missing_refs['dataset_name'])]\n",
    "\n",
    "# Adjust column names in missing_refs to match carlino_supp\n",
    "missing_refs = missing_refs.rename(columns={'dataset_name': 'Dataset', 'project_accession': 'Accession number'})\n",
    "\n",
    "\n",
    "# concatenate carlino_supp and missing_refs\n",
    "carlino_supp_updated = pd.concat([carlino_supp_updated, missing_refs], ignore_index=True)\n",
    "\n",
    "# remove rows missing references\n",
    "carlino_supp_updated = carlino_supp_updated.dropna(subset=['Reference'])\n",
    "\n",
    "# check if the number of rows is the same as the sum of the two dataframes minus the duplicates\n",
    "print(f\"Number of rows in the concatenated dataframe: {carlino_supp_updated.shape[0]}\")\n",
    "print(f\"Sum of the two dataframes: {carlino_supp.shape[0] + missing_refs.shape[0] - 4}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of rows with References: 3186\n",
      "Total rows in carlino_metadata: 3368\n"
     ]
    }
   ],
   "source": [
    "# Split rows with multiple bioproject accession numbers in carlino_supp\n",
    "carlino_supp_updated['Accession number'] = carlino_supp_updated['Accession number'].str.split(',')\n",
    "carlino_supp_updated = carlino_supp_updated.explode('Accession number')\n",
    "carlino_supp_updated['Accession number'] = carlino_supp_updated['Accession number'].str.strip()\n",
    "\n",
    "# Create a mapping dictionary using both Dataset and Accession number as keys\n",
    "reference_dict = {}\n",
    "for _, row in carlino_supp_updated.iterrows():\n",
    "    # Map by dataset name\n",
    "    reference_dict[('dataset', row['Dataset'])] = row['Reference']\n",
    "    # Map by accession if available\n",
    "    if pd.notna(row['Accession number']):\n",
    "        reference_dict[('accession', row['Accession number'])] = row['Reference']\n",
    "\n",
    "# Add Reference column to carlino_metadata using both dataset_name and project_accession\n",
    "def get_reference(row):\n",
    "    # Try to find reference by dataset name first\n",
    "    ref = reference_dict.get(('dataset', row['dataset_name']))\n",
    "    if pd.isna(ref) and pd.notna(row['project_accession']):\n",
    "        # If not found, try to find by project accession\n",
    "        ref = reference_dict.get(('accession', row['project_accession']))\n",
    "    return ref\n",
    "\n",
    "carlino_metadata['Reference'] = carlino_metadata.apply(get_reference, axis=1)\n",
    "\n",
    "# Verify the merge by checking if References were added\n",
    "print(\"Number of rows with References:\", carlino_metadata['Reference'].notna().sum())\n",
    "print(\"Total rows in carlino_metadata:\", len(carlino_metadata))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>dataset_name</th>\n",
       "      <th>project_accession</th>\n",
       "      <th>Reference</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [dataset_name, project_accession, Reference]\n",
       "Index: []"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# checking the missing references again\n",
    "missing_refs = carlino_metadata[\n",
    "    (carlino_metadata.Reference.isna()) & \n",
    "    (carlino_metadata['fermented/non-fermented'] == 'F')\n",
    "][['dataset_name', 'project_accession','Reference']].drop_duplicates()\n",
    "\n",
    "# check if the dataset_name is in the Fermentos metadata\n",
    "missing_refs = missing_refs[missing_refs['dataset_name'].apply(lambda x: dataset['mag_id'].str.contains(x).any())]\n",
    "display(missing_refs)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# addiing References is complete for carlino metadata. \n",
    "# I will save the table now to use it later. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "metadata": {},
   "outputs": [],
   "source": [
    "# save carlino_metadata to a file\n",
    "carlino_metadata.to_csv('data/cFMD_metadata_updated.tsv', sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Adjusting additional metadata sets. \n",
    "Since Carlino is by far the largest dataset we have, I will adjust the other smaller sets to match it as much as possible. <br>\n",
    "I will then concatenated all the added metadata information before merging it with the current metadata file. <br>\n",
    "Once all the additional data is added I will clean the data. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Run</th>\n",
       "      <th>Assay Type</th>\n",
       "      <th>AvgSpotLen</th>\n",
       "      <th>Bases</th>\n",
       "      <th>BioProject</th>\n",
       "      <th>BioSample</th>\n",
       "      <th>BioSampleModel</th>\n",
       "      <th>Bytes</th>\n",
       "      <th>Center Name</th>\n",
       "      <th>Collection_Date</th>\n",
       "      <th>...</th>\n",
       "      <th>num_replicons</th>\n",
       "      <th>Organism</th>\n",
       "      <th>Platform</th>\n",
       "      <th>ref_biomaterial</th>\n",
       "      <th>ReleaseDate</th>\n",
       "      <th>create_date</th>\n",
       "      <th>version</th>\n",
       "      <th>Sample Name</th>\n",
       "      <th>SRA Study</th>\n",
       "      <th>strain</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SRR19240301; SRR19240302</td>\n",
       "      <td>WGS</td>\n",
       "      <td>6350; 300</td>\n",
       "      <td>76630626; 820718700</td>\n",
       "      <td>PRJNA837789</td>\n",
       "      <td>SAMN28230942</td>\n",
       "      <td>MIGS.ba</td>\n",
       "      <td>67972364; 391048120</td>\n",
       "      <td>UNIVERSITY OF CALIFORNIA SAN DIEGO</td>\n",
       "      <td>2020-08-20</td>\n",
       "      <td>...</td>\n",
       "      <td>n.a.</td>\n",
       "      <td>Psychrobacter sp. CCS169</td>\n",
       "      <td>OXFORD_NANOPORE; ILLUMINA</td>\n",
       "      <td>n.a.</td>\n",
       "      <td>2022-11-01T00:00:00Z</td>\n",
       "      <td>2022-05-17T08:03:00Z</td>\n",
       "      <td>1</td>\n",
       "      <td>CCS169</td>\n",
       "      <td>SRP375596</td>\n",
       "      <td>CCS169</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SRR19240304; SRR19240305</td>\n",
       "      <td>WGS</td>\n",
       "      <td>6852; 300</td>\n",
       "      <td>169732241; 896587500</td>\n",
       "      <td>PRJNA838106</td>\n",
       "      <td>SAMN28404985</td>\n",
       "      <td>MIGS.ba</td>\n",
       "      <td>154338594; 444876504</td>\n",
       "      <td>UNIVERSITY OF CALIFORNIA SAN DIEGO</td>\n",
       "      <td>2020-08-20</td>\n",
       "      <td>...</td>\n",
       "      <td>n.a.</td>\n",
       "      <td>Brachybacterium sp. CCS183</td>\n",
       "      <td>OXFORD_NANOPORE; ILLUMINA</td>\n",
       "      <td>n.a.</td>\n",
       "      <td>2022-11-01T00:00:00Z</td>\n",
       "      <td>2022-05-17T08:07:00Z; 2022-05-17T08:04:00Z</td>\n",
       "      <td>1</td>\n",
       "      <td>CCS183</td>\n",
       "      <td>SRP375598</td>\n",
       "      <td>CCS183</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SRR19236863; SRR19236864</td>\n",
       "      <td>WGS</td>\n",
       "      <td>8410; 300</td>\n",
       "      <td>435143545; 846789000</td>\n",
       "      <td>PRJNA838091</td>\n",
       "      <td>SAMN28404911</td>\n",
       "      <td>MIGS.ba</td>\n",
       "      <td>377507949; 419527532</td>\n",
       "      <td>UNIVERSITY OF CALIFORNIA SAN DIEGO</td>\n",
       "      <td>2020-08-20</td>\n",
       "      <td>...</td>\n",
       "      <td>n.a.</td>\n",
       "      <td>Brevibacterium aurantiacum</td>\n",
       "      <td>OXFORD_NANOPORE; ILLUMINA</td>\n",
       "      <td>n.a.</td>\n",
       "      <td>2022-11-01T00:00:00Z</td>\n",
       "      <td>2022-05-16T18:28:00Z; 2022-05-16T18:23:00Z</td>\n",
       "      <td>1</td>\n",
       "      <td>CCS174</td>\n",
       "      <td>SRP375504</td>\n",
       "      <td>CCS174</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SRR19236806</td>\n",
       "      <td>WGS</td>\n",
       "      <td>300</td>\n",
       "      <td>993764400</td>\n",
       "      <td>PRJNA838095</td>\n",
       "      <td>SAMN28404915</td>\n",
       "      <td>MIGS.ba</td>\n",
       "      <td>450318221</td>\n",
       "      <td>UNIVERSITY OF CALIFORNIA SAN DIEGO</td>\n",
       "      <td>2020-08-20</td>\n",
       "      <td>...</td>\n",
       "      <td>n.a.</td>\n",
       "      <td>Staphylococcus carnosus</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>n.a.</td>\n",
       "      <td>2022-11-01T00:00:00Z</td>\n",
       "      <td>2022-05-16T18:22:00Z</td>\n",
       "      <td>1</td>\n",
       "      <td>CCS178</td>\n",
       "      <td>SRP375502</td>\n",
       "      <td>CCS178</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SRR19236869; SRR19236870</td>\n",
       "      <td>WGS</td>\n",
       "      <td>8852; 300</td>\n",
       "      <td>883564434; 1047163800</td>\n",
       "      <td>PRJNA837750</td>\n",
       "      <td>SAMN28230094</td>\n",
       "      <td>MIGS.ba</td>\n",
       "      <td>780487290; 518534252</td>\n",
       "      <td>UNIVERSITY OF CALIFORNIA SAN DIEGO</td>\n",
       "      <td>2020-08-20</td>\n",
       "      <td>...</td>\n",
       "      <td>n.a.</td>\n",
       "      <td>Brevibacterium aurantiacum</td>\n",
       "      <td>OXFORD_NANOPORE; ILLUMINA</td>\n",
       "      <td>n.a.</td>\n",
       "      <td>2022-11-01T00:00:00Z</td>\n",
       "      <td>2022-05-16T18:28:00Z; 2022-05-16T18:23:00Z</td>\n",
       "      <td>1</td>\n",
       "      <td>CCS156</td>\n",
       "      <td>SRP375506</td>\n",
       "      <td>CCS156</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>SRR19236865</td>\n",
       "      <td>WGS</td>\n",
       "      <td>300</td>\n",
       "      <td>885285600</td>\n",
       "      <td>PRJNA838094</td>\n",
       "      <td>SAMN28404914</td>\n",
       "      <td>MIGS.ba</td>\n",
       "      <td>404201531</td>\n",
       "      <td>UNIVERSITY OF CALIFORNIA SAN DIEGO</td>\n",
       "      <td>2020-08-20</td>\n",
       "      <td>...</td>\n",
       "      <td>n.a.</td>\n",
       "      <td>Staphylococcus xylosus</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>n.a.</td>\n",
       "      <td>2022-11-01T00:00:00Z</td>\n",
       "      <td>2022-05-16T18:23:00Z</td>\n",
       "      <td>1</td>\n",
       "      <td>CCS177</td>\n",
       "      <td>SRP375505</td>\n",
       "      <td>CCS177</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>SRR19240297; SRR19240298</td>\n",
       "      <td>WGS</td>\n",
       "      <td>6231; 300</td>\n",
       "      <td>825058471; 895976700</td>\n",
       "      <td>PRJNA838262</td>\n",
       "      <td>SAMN28416482</td>\n",
       "      <td>MIGS.ba</td>\n",
       "      <td>721809360; 425596627</td>\n",
       "      <td>UNIVERSITY OF CALIFORNIA SAN DIEGO</td>\n",
       "      <td>2020-08-20</td>\n",
       "      <td>...</td>\n",
       "      <td>n.a.</td>\n",
       "      <td>Psychrobacter sp. CCS184</td>\n",
       "      <td>OXFORD_NANOPORE; ILLUMINA</td>\n",
       "      <td>n.a.</td>\n",
       "      <td>2022-11-01T00:00:00Z</td>\n",
       "      <td>2022-05-17T08:05:00Z; 2022-05-17T08:03:00Z</td>\n",
       "      <td>1</td>\n",
       "      <td>CCS184</td>\n",
       "      <td>SRP375594</td>\n",
       "      <td>CCS184</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>SRR19240247; SRR19240248</td>\n",
       "      <td>WGS</td>\n",
       "      <td>9500; 300</td>\n",
       "      <td>1906080156; 1065953700</td>\n",
       "      <td>PRJNA837754</td>\n",
       "      <td>SAMN28230173</td>\n",
       "      <td>MIGS.ba</td>\n",
       "      <td>1707103809; 511862703</td>\n",
       "      <td>UNIVERSITY OF CALIFORNIA SAN DIEGO</td>\n",
       "      <td>2020-08-20</td>\n",
       "      <td>...</td>\n",
       "      <td>n.a.</td>\n",
       "      <td>Halomonas sp. CCS158</td>\n",
       "      <td>OXFORD_NANOPORE; ILLUMINA</td>\n",
       "      <td>n.a.</td>\n",
       "      <td>2022-11-01T00:00:00Z</td>\n",
       "      <td>2022-05-17T07:50:00Z; 2022-05-17T07:42:00Z</td>\n",
       "      <td>1</td>\n",
       "      <td>CCS158</td>\n",
       "      <td>SRP375591</td>\n",
       "      <td>CCS158</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>SRR19237397; SRR19237398</td>\n",
       "      <td>WGS</td>\n",
       "      <td>13050; 300</td>\n",
       "      <td>194158638; 962111100</td>\n",
       "      <td>PRJNA837770</td>\n",
       "      <td>SAMN28230657</td>\n",
       "      <td>MIGS.ba</td>\n",
       "      <td>169592768; 479456251</td>\n",
       "      <td>UNIVERSITY OF CALIFORNIA SAN DIEGO</td>\n",
       "      <td>2020-08-20</td>\n",
       "      <td>...</td>\n",
       "      <td>n.a.</td>\n",
       "      <td>Microbacterium gubbeenense</td>\n",
       "      <td>OXFORD_NANOPORE; ILLUMINA</td>\n",
       "      <td>n.a.</td>\n",
       "      <td>2022-11-01T00:00:00Z</td>\n",
       "      <td>2022-05-16T18:46:00Z; 2022-05-16T18:43:00Z</td>\n",
       "      <td>1</td>\n",
       "      <td>CCS160</td>\n",
       "      <td>SRP375516</td>\n",
       "      <td>CCS160</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>SRR19236881; SRR19236882</td>\n",
       "      <td>WGS</td>\n",
       "      <td>5601; 300</td>\n",
       "      <td>66742337; 812418300</td>\n",
       "      <td>PRJNA838100</td>\n",
       "      <td>SAMN28404984</td>\n",
       "      <td>MIGS.ba</td>\n",
       "      <td>59270012; 378102207</td>\n",
       "      <td>UNIVERSITY OF CALIFORNIA SAN DIEGO</td>\n",
       "      <td>2020-08-20</td>\n",
       "      <td>...</td>\n",
       "      <td>n.a.</td>\n",
       "      <td>Psychrobacter cibarius</td>\n",
       "      <td>OXFORD_NANOPORE; ILLUMINA</td>\n",
       "      <td>n.a.</td>\n",
       "      <td>2022-11-01T00:00:00Z</td>\n",
       "      <td>2022-05-16T18:23:00Z</td>\n",
       "      <td>1</td>\n",
       "      <td>CCS182</td>\n",
       "      <td>SRP375508</td>\n",
       "      <td>CCS182</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>SRR19240245; SRR19240246</td>\n",
       "      <td>WGS</td>\n",
       "      <td>300; 5886</td>\n",
       "      <td>865814100; 1194238663</td>\n",
       "      <td>PRJNA838104</td>\n",
       "      <td>SAMN28404982</td>\n",
       "      <td>MIGS.ba</td>\n",
       "      <td>408582509; 1069924971</td>\n",
       "      <td>UNIVERSITY OF CALIFORNIA SAN DIEGO</td>\n",
       "      <td>2020-08-20</td>\n",
       "      <td>...</td>\n",
       "      <td>n.a.</td>\n",
       "      <td>Psychrobacter sp. CCS180</td>\n",
       "      <td>ILLUMINA; OXFORD_NANOPORE</td>\n",
       "      <td>n.a.</td>\n",
       "      <td>2022-11-01T00:00:00Z</td>\n",
       "      <td>2022-05-17T07:42:00Z; 2022-05-17T07:45:00Z</td>\n",
       "      <td>1</td>\n",
       "      <td>CCS180</td>\n",
       "      <td>SRP375590</td>\n",
       "      <td>CCS180</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>SRR19240303</td>\n",
       "      <td>WGS</td>\n",
       "      <td>300</td>\n",
       "      <td>789764400</td>\n",
       "      <td>PRJNA838092</td>\n",
       "      <td>SAMN28404912</td>\n",
       "      <td>MIGS.ba</td>\n",
       "      <td>376068475</td>\n",
       "      <td>UNIVERSITY OF CALIFORNIA SAN DIEGO</td>\n",
       "      <td>2020-08-20</td>\n",
       "      <td>...</td>\n",
       "      <td>n.a.</td>\n",
       "      <td>Psychrobacter sp. CCS175</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>n.a.</td>\n",
       "      <td>2022-11-01T00:00:00Z</td>\n",
       "      <td>2022-05-17T08:04:00Z</td>\n",
       "      <td>1</td>\n",
       "      <td>CCS175</td>\n",
       "      <td>SRP375597</td>\n",
       "      <td>CCS175</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>SRR19236887; SRR19236888</td>\n",
       "      <td>WGS</td>\n",
       "      <td>9503; 300</td>\n",
       "      <td>519463050; 875304600</td>\n",
       "      <td>PRJNA838105</td>\n",
       "      <td>SAMN28404981</td>\n",
       "      <td>MIGS.ba</td>\n",
       "      <td>457561510; 408408946</td>\n",
       "      <td>UNIVERSITY OF CALIFORNIA SAN DIEGO</td>\n",
       "      <td>2020-08-20</td>\n",
       "      <td>...</td>\n",
       "      <td>n.a.</td>\n",
       "      <td>Psychrobacter immobilis</td>\n",
       "      <td>OXFORD_NANOPORE; ILLUMINA</td>\n",
       "      <td>n.a.</td>\n",
       "      <td>2022-11-01T00:00:00Z</td>\n",
       "      <td>2022-05-16T18:29:00Z; 2022-05-16T18:24:00Z</td>\n",
       "      <td>1</td>\n",
       "      <td>CCS179</td>\n",
       "      <td>SRP375510</td>\n",
       "      <td>CCS179</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>SRR19236804; SRR19236805</td>\n",
       "      <td>WGS</td>\n",
       "      <td>12734; 300</td>\n",
       "      <td>88428573; 697548600</td>\n",
       "      <td>PRJNA837776</td>\n",
       "      <td>SAMN28230850</td>\n",
       "      <td>MIGS.ba</td>\n",
       "      <td>78166994; 336561572</td>\n",
       "      <td>UNIVERSITY OF CALIFORNIA SAN DIEGO</td>\n",
       "      <td>2020-08-20</td>\n",
       "      <td>...</td>\n",
       "      <td>n.a.</td>\n",
       "      <td>Psychrobacter alimentarius</td>\n",
       "      <td>OXFORD_NANOPORE; ILLUMINA</td>\n",
       "      <td>n.a.</td>\n",
       "      <td>2022-11-01T00:00:00Z</td>\n",
       "      <td>2022-05-16T18:22:00Z</td>\n",
       "      <td>1</td>\n",
       "      <td>CCS164</td>\n",
       "      <td>SRP375501</td>\n",
       "      <td>CCS164</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>SRR19236885; SRR19236886</td>\n",
       "      <td>WGS</td>\n",
       "      <td>4955; 300</td>\n",
       "      <td>598601405; 839253900</td>\n",
       "      <td>PRJNA838093</td>\n",
       "      <td>SAMN28404913</td>\n",
       "      <td>MIGS.ba</td>\n",
       "      <td>515484009; 391768040</td>\n",
       "      <td>UNIVERSITY OF CALIFORNIA SAN DIEGO</td>\n",
       "      <td>2020-08-20</td>\n",
       "      <td>...</td>\n",
       "      <td>n.a.</td>\n",
       "      <td>Pseudoalteromonas nigrifaciens</td>\n",
       "      <td>OXFORD_NANOPORE; ILLUMINA</td>\n",
       "      <td>n.a.</td>\n",
       "      <td>2022-11-01T00:00:00Z</td>\n",
       "      <td>2022-05-16T18:25:00Z; 2022-05-16T18:24:00Z</td>\n",
       "      <td>1</td>\n",
       "      <td>CCS176</td>\n",
       "      <td>SRP375509</td>\n",
       "      <td>CCS176</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>SRR19240306; SRR19240307</td>\n",
       "      <td>WGS</td>\n",
       "      <td>5606; 300</td>\n",
       "      <td>1270904742; 932733600</td>\n",
       "      <td>PRJNA838102</td>\n",
       "      <td>SAMN28404983</td>\n",
       "      <td>MIGS.ba</td>\n",
       "      <td>1139715534; 436798278</td>\n",
       "      <td>UNIVERSITY OF CALIFORNIA SAN DIEGO</td>\n",
       "      <td>2020-08-20</td>\n",
       "      <td>...</td>\n",
       "      <td>n.a.</td>\n",
       "      <td>Psychrobacter sp. CCS181</td>\n",
       "      <td>OXFORD_NANOPORE; ILLUMINA</td>\n",
       "      <td>n.a.</td>\n",
       "      <td>2022-11-01T00:00:00Z</td>\n",
       "      <td>2022-05-17T08:07:00Z; 2022-05-17T08:04:00Z</td>\n",
       "      <td>1</td>\n",
       "      <td>CCS181</td>\n",
       "      <td>SRP375599</td>\n",
       "      <td>CCS181</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>16 rows × 38 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                         Run Assay Type  AvgSpotLen                   Bases  \\\n",
       "0   SRR19240301; SRR19240302        WGS   6350; 300     76630626; 820718700   \n",
       "1   SRR19240304; SRR19240305        WGS   6852; 300    169732241; 896587500   \n",
       "2   SRR19236863; SRR19236864        WGS   8410; 300    435143545; 846789000   \n",
       "3                SRR19236806        WGS         300               993764400   \n",
       "4   SRR19236869; SRR19236870        WGS   8852; 300   883564434; 1047163800   \n",
       "5                SRR19236865        WGS         300               885285600   \n",
       "6   SRR19240297; SRR19240298        WGS   6231; 300    825058471; 895976700   \n",
       "7   SRR19240247; SRR19240248        WGS   9500; 300  1906080156; 1065953700   \n",
       "8   SRR19237397; SRR19237398        WGS  13050; 300    194158638; 962111100   \n",
       "9   SRR19236881; SRR19236882        WGS   5601; 300     66742337; 812418300   \n",
       "10  SRR19240245; SRR19240246        WGS   300; 5886   865814100; 1194238663   \n",
       "11               SRR19240303        WGS         300               789764400   \n",
       "12  SRR19236887; SRR19236888        WGS   9503; 300    519463050; 875304600   \n",
       "13  SRR19236804; SRR19236805        WGS  12734; 300     88428573; 697548600   \n",
       "14  SRR19236885; SRR19236886        WGS   4955; 300    598601405; 839253900   \n",
       "15  SRR19240306; SRR19240307        WGS   5606; 300   1270904742; 932733600   \n",
       "\n",
       "     BioProject     BioSample BioSampleModel                  Bytes  \\\n",
       "0   PRJNA837789  SAMN28230942        MIGS.ba    67972364; 391048120   \n",
       "1   PRJNA838106  SAMN28404985        MIGS.ba   154338594; 444876504   \n",
       "2   PRJNA838091  SAMN28404911        MIGS.ba   377507949; 419527532   \n",
       "3   PRJNA838095  SAMN28404915        MIGS.ba              450318221   \n",
       "4   PRJNA837750  SAMN28230094        MIGS.ba   780487290; 518534252   \n",
       "5   PRJNA838094  SAMN28404914        MIGS.ba              404201531   \n",
       "6   PRJNA838262  SAMN28416482        MIGS.ba   721809360; 425596627   \n",
       "7   PRJNA837754  SAMN28230173        MIGS.ba  1707103809; 511862703   \n",
       "8   PRJNA837770  SAMN28230657        MIGS.ba   169592768; 479456251   \n",
       "9   PRJNA838100  SAMN28404984        MIGS.ba    59270012; 378102207   \n",
       "10  PRJNA838104  SAMN28404982        MIGS.ba  408582509; 1069924971   \n",
       "11  PRJNA838092  SAMN28404912        MIGS.ba              376068475   \n",
       "12  PRJNA838105  SAMN28404981        MIGS.ba   457561510; 408408946   \n",
       "13  PRJNA837776  SAMN28230850        MIGS.ba    78166994; 336561572   \n",
       "14  PRJNA838093  SAMN28404913        MIGS.ba   515484009; 391768040   \n",
       "15  PRJNA838102  SAMN28404983        MIGS.ba  1139715534; 436798278   \n",
       "\n",
       "                           Center Name Collection_Date  ... num_replicons  \\\n",
       "0   UNIVERSITY OF CALIFORNIA SAN DIEGO      2020-08-20  ...          n.a.   \n",
       "1   UNIVERSITY OF CALIFORNIA SAN DIEGO      2020-08-20  ...          n.a.   \n",
       "2   UNIVERSITY OF CALIFORNIA SAN DIEGO      2020-08-20  ...          n.a.   \n",
       "3   UNIVERSITY OF CALIFORNIA SAN DIEGO      2020-08-20  ...          n.a.   \n",
       "4   UNIVERSITY OF CALIFORNIA SAN DIEGO      2020-08-20  ...          n.a.   \n",
       "5   UNIVERSITY OF CALIFORNIA SAN DIEGO      2020-08-20  ...          n.a.   \n",
       "6   UNIVERSITY OF CALIFORNIA SAN DIEGO      2020-08-20  ...          n.a.   \n",
       "7   UNIVERSITY OF CALIFORNIA SAN DIEGO      2020-08-20  ...          n.a.   \n",
       "8   UNIVERSITY OF CALIFORNIA SAN DIEGO      2020-08-20  ...          n.a.   \n",
       "9   UNIVERSITY OF CALIFORNIA SAN DIEGO      2020-08-20  ...          n.a.   \n",
       "10  UNIVERSITY OF CALIFORNIA SAN DIEGO      2020-08-20  ...          n.a.   \n",
       "11  UNIVERSITY OF CALIFORNIA SAN DIEGO      2020-08-20  ...          n.a.   \n",
       "12  UNIVERSITY OF CALIFORNIA SAN DIEGO      2020-08-20  ...          n.a.   \n",
       "13  UNIVERSITY OF CALIFORNIA SAN DIEGO      2020-08-20  ...          n.a.   \n",
       "14  UNIVERSITY OF CALIFORNIA SAN DIEGO      2020-08-20  ...          n.a.   \n",
       "15  UNIVERSITY OF CALIFORNIA SAN DIEGO      2020-08-20  ...          n.a.   \n",
       "\n",
       "                          Organism                   Platform ref_biomaterial  \\\n",
       "0         Psychrobacter sp. CCS169  OXFORD_NANOPORE; ILLUMINA            n.a.   \n",
       "1       Brachybacterium sp. CCS183  OXFORD_NANOPORE; ILLUMINA            n.a.   \n",
       "2       Brevibacterium aurantiacum  OXFORD_NANOPORE; ILLUMINA            n.a.   \n",
       "3          Staphylococcus carnosus                   ILLUMINA            n.a.   \n",
       "4       Brevibacterium aurantiacum  OXFORD_NANOPORE; ILLUMINA            n.a.   \n",
       "5           Staphylococcus xylosus                   ILLUMINA            n.a.   \n",
       "6         Psychrobacter sp. CCS184  OXFORD_NANOPORE; ILLUMINA            n.a.   \n",
       "7             Halomonas sp. CCS158  OXFORD_NANOPORE; ILLUMINA            n.a.   \n",
       "8       Microbacterium gubbeenense  OXFORD_NANOPORE; ILLUMINA            n.a.   \n",
       "9           Psychrobacter cibarius  OXFORD_NANOPORE; ILLUMINA            n.a.   \n",
       "10        Psychrobacter sp. CCS180  ILLUMINA; OXFORD_NANOPORE            n.a.   \n",
       "11        Psychrobacter sp. CCS175                   ILLUMINA            n.a.   \n",
       "12         Psychrobacter immobilis  OXFORD_NANOPORE; ILLUMINA            n.a.   \n",
       "13      Psychrobacter alimentarius  OXFORD_NANOPORE; ILLUMINA            n.a.   \n",
       "14  Pseudoalteromonas nigrifaciens  OXFORD_NANOPORE; ILLUMINA            n.a.   \n",
       "15        Psychrobacter sp. CCS181  OXFORD_NANOPORE; ILLUMINA            n.a.   \n",
       "\n",
       "             ReleaseDate                                 create_date version  \\\n",
       "0   2022-11-01T00:00:00Z                        2022-05-17T08:03:00Z       1   \n",
       "1   2022-11-01T00:00:00Z  2022-05-17T08:07:00Z; 2022-05-17T08:04:00Z       1   \n",
       "2   2022-11-01T00:00:00Z  2022-05-16T18:28:00Z; 2022-05-16T18:23:00Z       1   \n",
       "3   2022-11-01T00:00:00Z                        2022-05-16T18:22:00Z       1   \n",
       "4   2022-11-01T00:00:00Z  2022-05-16T18:28:00Z; 2022-05-16T18:23:00Z       1   \n",
       "5   2022-11-01T00:00:00Z                        2022-05-16T18:23:00Z       1   \n",
       "6   2022-11-01T00:00:00Z  2022-05-17T08:05:00Z; 2022-05-17T08:03:00Z       1   \n",
       "7   2022-11-01T00:00:00Z  2022-05-17T07:50:00Z; 2022-05-17T07:42:00Z       1   \n",
       "8   2022-11-01T00:00:00Z  2022-05-16T18:46:00Z; 2022-05-16T18:43:00Z       1   \n",
       "9   2022-11-01T00:00:00Z                        2022-05-16T18:23:00Z       1   \n",
       "10  2022-11-01T00:00:00Z  2022-05-17T07:42:00Z; 2022-05-17T07:45:00Z       1   \n",
       "11  2022-11-01T00:00:00Z                        2022-05-17T08:04:00Z       1   \n",
       "12  2022-11-01T00:00:00Z  2022-05-16T18:29:00Z; 2022-05-16T18:24:00Z       1   \n",
       "13  2022-11-01T00:00:00Z                        2022-05-16T18:22:00Z       1   \n",
       "14  2022-11-01T00:00:00Z  2022-05-16T18:25:00Z; 2022-05-16T18:24:00Z       1   \n",
       "15  2022-11-01T00:00:00Z  2022-05-17T08:07:00Z; 2022-05-17T08:04:00Z       1   \n",
       "\n",
       "   Sample Name  SRA Study  strain  \n",
       "0       CCS169  SRP375596  CCS169  \n",
       "1       CCS183  SRP375598  CCS183  \n",
       "2       CCS174  SRP375504  CCS174  \n",
       "3       CCS178  SRP375502  CCS178  \n",
       "4       CCS156  SRP375506  CCS156  \n",
       "5       CCS177  SRP375505  CCS177  \n",
       "6       CCS184  SRP375594  CCS184  \n",
       "7       CCS158  SRP375591  CCS158  \n",
       "8       CCS160  SRP375516  CCS160  \n",
       "9       CCS182  SRP375508  CCS182  \n",
       "10      CCS180  SRP375590  CCS180  \n",
       "11      CCS175  SRP375597  CCS175  \n",
       "12      CCS179  SRP375510  CCS179  \n",
       "13      CCS164  SRP375501  CCS164  \n",
       "14      CCS176  SRP375509  CCS176  \n",
       "15      CCS181  SRP375599  CCS181  \n",
       "\n",
       "[16 rows x 38 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "import glob\n",
    "\n",
    "# Get all files matching the pattern\n",
    "files = glob.glob('data/Saak2023_CheeseMAGs_bin*')\n",
    "\n",
    "# Function to merge two rows into one\n",
    "def merge_rows(row1, row2):\n",
    "    merged = {}\n",
    "    for col in row1.index:\n",
    "        # If values are the same, keep one\n",
    "        if row1[col] == row2[col]:\n",
    "            merged[col] = row1[col]\n",
    "        # If values are different and not NaN, combine them\n",
    "        elif pd.notna(row1[col]) and pd.notna(row2[col]):\n",
    "            merged[col] = f\"{row1[col]}; {row2[col]}\"\n",
    "        # If one value is NaN, keep the non-NaN value\n",
    "        else:\n",
    "            merged[col] = row1[col] if pd.notna(row1[col]) else row2[col]\n",
    "    return pd.Series(merged)\n",
    "\n",
    "# List to store all combined dataframes\n",
    "all_dfs = []\n",
    "\n",
    "# Process each file\n",
    "for file in files:\n",
    "    # Read the file\n",
    "    df = pd.read_csv(file)\n",
    "    \n",
    "    # If there's only one row, add it directly\n",
    "    if len(df) == 1:\n",
    "        all_dfs.append(df)\n",
    "    # If there are multiple rows, merge them\n",
    "    else:\n",
    "        merged_row = df.iloc[0]\n",
    "        for i in range(1, len(df)):\n",
    "            merged_row = merge_rows(merged_row, df.iloc[i])\n",
    "        all_dfs.append(pd.DataFrame([merged_row]))\n",
    "\n",
    "# Combine all dataframes\n",
    "Saak2023_metadata = pd.concat(all_dfs, ignore_index=True)\n",
    "\n",
    "# Display result\n",
    "display(Saak2023_metadata)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Run</th>\n",
       "      <th>Assay Type</th>\n",
       "      <th>AvgSpotLen</th>\n",
       "      <th>Bases</th>\n",
       "      <th>batch</th>\n",
       "      <th>BioProject</th>\n",
       "      <th>BioSample</th>\n",
       "      <th>BioSampleModel</th>\n",
       "      <th>Bytes</th>\n",
       "      <th>Center Name</th>\n",
       "      <th>...</th>\n",
       "      <th>LibrarySelection</th>\n",
       "      <th>LibrarySource</th>\n",
       "      <th>Organism</th>\n",
       "      <th>Platform</th>\n",
       "      <th>ReleaseDate</th>\n",
       "      <th>create_date</th>\n",
       "      <th>version</th>\n",
       "      <th>Sample Name</th>\n",
       "      <th>SRA Study</th>\n",
       "      <th>week</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SRR16955650</td>\n",
       "      <td>AMPLICON</td>\n",
       "      <td>302</td>\n",
       "      <td>15392034</td>\n",
       "      <td>2</td>\n",
       "      <td>PRJNA778418</td>\n",
       "      <td>SAMN23169395</td>\n",
       "      <td>MIMS.me,MIGS/MIMS/MIMARKS.microbial</td>\n",
       "      <td>4849268</td>\n",
       "      <td>UNIVERSITY OF CALIFORNIA SAN DIEGO</td>\n",
       "      <td>...</td>\n",
       "      <td>PCR</td>\n",
       "      <td>METAGENOMIC</td>\n",
       "      <td>food fermentation metagenome</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>2022-11-01T00:00:00Z</td>\n",
       "      <td>2021-11-16T10:35:00Z</td>\n",
       "      <td>1</td>\n",
       "      <td>O_2_4_CheeseA</td>\n",
       "      <td>SRP346349</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SRR16955651</td>\n",
       "      <td>AMPLICON</td>\n",
       "      <td>302</td>\n",
       "      <td>16215286</td>\n",
       "      <td>2</td>\n",
       "      <td>PRJNA778418</td>\n",
       "      <td>SAMN23169394</td>\n",
       "      <td>MIMS.me,MIGS/MIMS/MIMARKS.microbial</td>\n",
       "      <td>5116627</td>\n",
       "      <td>UNIVERSITY OF CALIFORNIA SAN DIEGO</td>\n",
       "      <td>...</td>\n",
       "      <td>PCR</td>\n",
       "      <td>METAGENOMIC</td>\n",
       "      <td>food fermentation metagenome</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>2022-11-01T00:00:00Z</td>\n",
       "      <td>2021-11-16T10:35:00Z</td>\n",
       "      <td>1</td>\n",
       "      <td>O_2_3_CheeseA</td>\n",
       "      <td>SRP346349</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SRR16955652</td>\n",
       "      <td>AMPLICON</td>\n",
       "      <td>302</td>\n",
       "      <td>13833110</td>\n",
       "      <td>2</td>\n",
       "      <td>PRJNA778418</td>\n",
       "      <td>SAMN23169393</td>\n",
       "      <td>MIMS.me,MIGS/MIMS/MIMARKS.microbial</td>\n",
       "      <td>4415499</td>\n",
       "      <td>UNIVERSITY OF CALIFORNIA SAN DIEGO</td>\n",
       "      <td>...</td>\n",
       "      <td>PCR</td>\n",
       "      <td>METAGENOMIC</td>\n",
       "      <td>food fermentation metagenome</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>2022-11-01T00:00:00Z</td>\n",
       "      <td>2021-11-16T10:35:00Z</td>\n",
       "      <td>1</td>\n",
       "      <td>O_2_2_CheeseA</td>\n",
       "      <td>SRP346349</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SRR16955653</td>\n",
       "      <td>AMPLICON</td>\n",
       "      <td>302</td>\n",
       "      <td>15773460</td>\n",
       "      <td>2</td>\n",
       "      <td>PRJNA778418</td>\n",
       "      <td>SAMN23169392</td>\n",
       "      <td>MIMS.me,MIGS/MIMS/MIMARKS.microbial</td>\n",
       "      <td>5019625</td>\n",
       "      <td>UNIVERSITY OF CALIFORNIA SAN DIEGO</td>\n",
       "      <td>...</td>\n",
       "      <td>PCR</td>\n",
       "      <td>METAGENOMIC</td>\n",
       "      <td>food fermentation metagenome</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>2022-11-01T00:00:00Z</td>\n",
       "      <td>2021-11-16T10:35:00Z</td>\n",
       "      <td>1</td>\n",
       "      <td>O_2_1_CheeseA</td>\n",
       "      <td>SRP346349</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SRR16955654</td>\n",
       "      <td>AMPLICON</td>\n",
       "      <td>302</td>\n",
       "      <td>11930812</td>\n",
       "      <td>3</td>\n",
       "      <td>PRJNA778418</td>\n",
       "      <td>SAMN23169439</td>\n",
       "      <td>MIMS.me,MIGS/MIMS/MIMARKS.microbial</td>\n",
       "      <td>3850589</td>\n",
       "      <td>UNIVERSITY OF CALIFORNIA SAN DIEGO</td>\n",
       "      <td>...</td>\n",
       "      <td>PCR</td>\n",
       "      <td>METAGENOMIC</td>\n",
       "      <td>food fermentation metagenome</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>2022-11-01T00:00:00Z</td>\n",
       "      <td>2021-11-16T10:35:00Z</td>\n",
       "      <td>1</td>\n",
       "      <td>We_3_6_CheeseC</td>\n",
       "      <td>SRP346349</td>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 39 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           Run Assay Type  AvgSpotLen     Bases  batch   BioProject  \\\n",
       "0  SRR16955650   AMPLICON         302  15392034      2  PRJNA778418   \n",
       "1  SRR16955651   AMPLICON         302  16215286      2  PRJNA778418   \n",
       "2  SRR16955652   AMPLICON         302  13833110      2  PRJNA778418   \n",
       "3  SRR16955653   AMPLICON         302  15773460      2  PRJNA778418   \n",
       "4  SRR16955654   AMPLICON         302  11930812      3  PRJNA778418   \n",
       "\n",
       "      BioSample                       BioSampleModel    Bytes  \\\n",
       "0  SAMN23169395  MIMS.me,MIGS/MIMS/MIMARKS.microbial  4849268   \n",
       "1  SAMN23169394  MIMS.me,MIGS/MIMS/MIMARKS.microbial  5116627   \n",
       "2  SAMN23169393  MIMS.me,MIGS/MIMS/MIMARKS.microbial  4415499   \n",
       "3  SAMN23169392  MIMS.me,MIGS/MIMS/MIMARKS.microbial  5019625   \n",
       "4  SAMN23169439  MIMS.me,MIGS/MIMS/MIMARKS.microbial  3850589   \n",
       "\n",
       "                          Center Name  ... LibrarySelection LibrarySource  \\\n",
       "0  UNIVERSITY OF CALIFORNIA SAN DIEGO  ...              PCR   METAGENOMIC   \n",
       "1  UNIVERSITY OF CALIFORNIA SAN DIEGO  ...              PCR   METAGENOMIC   \n",
       "2  UNIVERSITY OF CALIFORNIA SAN DIEGO  ...              PCR   METAGENOMIC   \n",
       "3  UNIVERSITY OF CALIFORNIA SAN DIEGO  ...              PCR   METAGENOMIC   \n",
       "4  UNIVERSITY OF CALIFORNIA SAN DIEGO  ...              PCR   METAGENOMIC   \n",
       "\n",
       "                       Organism  Platform           ReleaseDate  \\\n",
       "0  food fermentation metagenome  ILLUMINA  2022-11-01T00:00:00Z   \n",
       "1  food fermentation metagenome  ILLUMINA  2022-11-01T00:00:00Z   \n",
       "2  food fermentation metagenome  ILLUMINA  2022-11-01T00:00:00Z   \n",
       "3  food fermentation metagenome  ILLUMINA  2022-11-01T00:00:00Z   \n",
       "4  food fermentation metagenome  ILLUMINA  2022-11-01T00:00:00Z   \n",
       "\n",
       "            create_date version     Sample Name  SRA Study week  \n",
       "0  2021-11-16T10:35:00Z       1   O_2_4_CheeseA  SRP346349    4  \n",
       "1  2021-11-16T10:35:00Z       1   O_2_3_CheeseA  SRP346349    3  \n",
       "2  2021-11-16T10:35:00Z       1   O_2_2_CheeseA  SRP346349    2  \n",
       "3  2021-11-16T10:35:00Z       1   O_2_1_CheeseA  SRP346349    1  \n",
       "4  2021-11-16T10:35:00Z       1  We_3_6_CheeseC  SRP346349   13  \n",
       "\n",
       "[5 rows x 39 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# upload the Saak2023 metagenome data\n",
    "Saak2023_metagenome = pd.read_csv('data/Saak2023_CheeseMAGs_PRJNA778418.csv')\n",
    "display(Saak2023_metagenome.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Columns in Saak2023_metadata: Index(['Run', 'Assay Type', 'AvgSpotLen', 'Bases', 'BioProject', 'BioSample',\n",
      "       'BioSampleModel', 'Bytes', 'Center Name', 'Collection_Date', 'Consent',\n",
      "       'DATASTORE filetype', 'DATASTORE provider', 'DATASTORE region',\n",
      "       'env_broad_scale', 'env_local_scale', 'env_medium', 'Experiment',\n",
      "       'geo_loc_name_country', 'geo_loc_name_country_continent',\n",
      "       'geo_loc_name', 'Instrument', 'isol_growth_condt', 'lat_lon',\n",
      "       'Library Name', 'LibraryLayout', 'LibrarySelection', 'LibrarySource',\n",
      "       'num_replicons', 'Organism', 'Platform', 'ref_biomaterial',\n",
      "       'ReleaseDate', 'create_date', 'version', 'Sample Name', 'SRA Study',\n",
      "       'strain'],\n",
      "      dtype='object')\n",
      "Columns in Saak2023_metagenome: Index(['Run', 'Assay Type', 'AvgSpotLen', 'Bases', 'batch', 'BioProject',\n",
      "       'BioSample', 'BioSampleModel', 'Bytes', 'Center Name', 'Cheese',\n",
      "       'Collection_Date', 'Consent', 'DATASTORE filetype',\n",
      "       'DATASTORE provider', 'DATASTORE region', 'Depth', 'elev',\n",
      "       'env_broad_scale', 'env_local_scale', 'env_medium', 'Experiment',\n",
      "       'geo_loc_name_country', 'geo_loc_name_country_continent',\n",
      "       'geo_loc_name', 'Instrument', 'lat_lon', 'Library Name',\n",
      "       'LibraryLayout', 'LibrarySelection', 'LibrarySource', 'Organism',\n",
      "       'Platform', 'ReleaseDate', 'create_date', 'version', 'Sample Name',\n",
      "       'SRA Study', 'week'],\n",
      "      dtype='object')\n",
      "\n",
      "Columns only in Saak2023_metadata:\n",
      "{'ref_biomaterial', 'isol_growth_condt', 'num_replicons', 'strain'}\n",
      "\n",
      "Columns only in Saak2023_metagenome:\n",
      "{'elev', 'Cheese', 'week', 'Depth', 'batch'}\n"
     ]
    }
   ],
   "source": [
    "# compare columns of Saak2023_metadata and Saak2023_metagenome\n",
    "print(\"Columns in Saak2023_metadata:\", Saak2023_metadata.columns)\n",
    "print(\"Columns in Saak2023_metagenome:\", Saak2023_metagenome.columns)\n",
    "\n",
    "# find the different columns and which table they are in\n",
    "# Print which columns are unique to each dataset\n",
    "saak_only = set(Saak2023_metadata.columns) - set(Saak2023_metagenome.columns)\n",
    "metagenome_only = set(c.columns) - set(Saak2023_metadata.columns)\n",
    "\n",
    "print(\"\\nColumns only in Saak2023_metadata:\")\n",
    "print(saak_only)\n",
    "print(\"\\nColumns only in Saak2023_metagenome:\")\n",
    "print(metagenome_only)\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Based on Fermentos metadata all the isolates are from cheese B so I can add that column. <br> \n",
    "I will also change the values in the metagenome to match the values in Fermentos metadata. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Run</th>\n",
       "      <th>Assay Type</th>\n",
       "      <th>AvgSpotLen</th>\n",
       "      <th>Bases</th>\n",
       "      <th>BioProject</th>\n",
       "      <th>BioSample</th>\n",
       "      <th>BioSampleModel</th>\n",
       "      <th>Bytes</th>\n",
       "      <th>Center Name</th>\n",
       "      <th>Collection_Date</th>\n",
       "      <th>...</th>\n",
       "      <th>create_date</th>\n",
       "      <th>version</th>\n",
       "      <th>Sample Name</th>\n",
       "      <th>SRA Study</th>\n",
       "      <th>strain</th>\n",
       "      <th>Cheese</th>\n",
       "      <th>batch</th>\n",
       "      <th>Depth</th>\n",
       "      <th>elev</th>\n",
       "      <th>week</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SRR19240301; SRR19240302</td>\n",
       "      <td>WGS</td>\n",
       "      <td>6350; 300</td>\n",
       "      <td>76630626; 820718700</td>\n",
       "      <td>PRJNA837789</td>\n",
       "      <td>SAMN28230942</td>\n",
       "      <td>MIGS.ba</td>\n",
       "      <td>67972364; 391048120</td>\n",
       "      <td>UNIVERSITY OF CALIFORNIA SAN DIEGO</td>\n",
       "      <td>2020-08-20</td>\n",
       "      <td>...</td>\n",
       "      <td>2022-05-17T08:03:00Z</td>\n",
       "      <td>1</td>\n",
       "      <td>CCS169</td>\n",
       "      <td>SRP375596</td>\n",
       "      <td>CCS169</td>\n",
       "      <td>cheese_B</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SRR19240304; SRR19240305</td>\n",
       "      <td>WGS</td>\n",
       "      <td>6852; 300</td>\n",
       "      <td>169732241; 896587500</td>\n",
       "      <td>PRJNA838106</td>\n",
       "      <td>SAMN28404985</td>\n",
       "      <td>MIGS.ba</td>\n",
       "      <td>154338594; 444876504</td>\n",
       "      <td>UNIVERSITY OF CALIFORNIA SAN DIEGO</td>\n",
       "      <td>2020-08-20</td>\n",
       "      <td>...</td>\n",
       "      <td>2022-05-17T08:07:00Z; 2022-05-17T08:04:00Z</td>\n",
       "      <td>1</td>\n",
       "      <td>CCS183</td>\n",
       "      <td>SRP375598</td>\n",
       "      <td>CCS183</td>\n",
       "      <td>cheese_B</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SRR19236863; SRR19236864</td>\n",
       "      <td>WGS</td>\n",
       "      <td>8410; 300</td>\n",
       "      <td>435143545; 846789000</td>\n",
       "      <td>PRJNA838091</td>\n",
       "      <td>SAMN28404911</td>\n",
       "      <td>MIGS.ba</td>\n",
       "      <td>377507949; 419527532</td>\n",
       "      <td>UNIVERSITY OF CALIFORNIA SAN DIEGO</td>\n",
       "      <td>2020-08-20</td>\n",
       "      <td>...</td>\n",
       "      <td>2022-05-16T18:28:00Z; 2022-05-16T18:23:00Z</td>\n",
       "      <td>1</td>\n",
       "      <td>CCS174</td>\n",
       "      <td>SRP375504</td>\n",
       "      <td>CCS174</td>\n",
       "      <td>cheese_B</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SRR19236806</td>\n",
       "      <td>WGS</td>\n",
       "      <td>300</td>\n",
       "      <td>993764400</td>\n",
       "      <td>PRJNA838095</td>\n",
       "      <td>SAMN28404915</td>\n",
       "      <td>MIGS.ba</td>\n",
       "      <td>450318221</td>\n",
       "      <td>UNIVERSITY OF CALIFORNIA SAN DIEGO</td>\n",
       "      <td>2020-08-20</td>\n",
       "      <td>...</td>\n",
       "      <td>2022-05-16T18:22:00Z</td>\n",
       "      <td>1</td>\n",
       "      <td>CCS178</td>\n",
       "      <td>SRP375502</td>\n",
       "      <td>CCS178</td>\n",
       "      <td>cheese_B</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SRR19236869; SRR19236870</td>\n",
       "      <td>WGS</td>\n",
       "      <td>8852; 300</td>\n",
       "      <td>883564434; 1047163800</td>\n",
       "      <td>PRJNA837750</td>\n",
       "      <td>SAMN28230094</td>\n",
       "      <td>MIGS.ba</td>\n",
       "      <td>780487290; 518534252</td>\n",
       "      <td>UNIVERSITY OF CALIFORNIA SAN DIEGO</td>\n",
       "      <td>2020-08-20</td>\n",
       "      <td>...</td>\n",
       "      <td>2022-05-16T18:28:00Z; 2022-05-16T18:23:00Z</td>\n",
       "      <td>1</td>\n",
       "      <td>CCS156</td>\n",
       "      <td>SRP375506</td>\n",
       "      <td>CCS156</td>\n",
       "      <td>cheese_B</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 43 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                        Run Assay Type AvgSpotLen                  Bases  \\\n",
       "0  SRR19240301; SRR19240302        WGS  6350; 300    76630626; 820718700   \n",
       "1  SRR19240304; SRR19240305        WGS  6852; 300   169732241; 896587500   \n",
       "2  SRR19236863; SRR19236864        WGS  8410; 300   435143545; 846789000   \n",
       "3               SRR19236806        WGS        300              993764400   \n",
       "4  SRR19236869; SRR19236870        WGS  8852; 300  883564434; 1047163800   \n",
       "\n",
       "    BioProject     BioSample BioSampleModel                 Bytes  \\\n",
       "0  PRJNA837789  SAMN28230942        MIGS.ba   67972364; 391048120   \n",
       "1  PRJNA838106  SAMN28404985        MIGS.ba  154338594; 444876504   \n",
       "2  PRJNA838091  SAMN28404911        MIGS.ba  377507949; 419527532   \n",
       "3  PRJNA838095  SAMN28404915        MIGS.ba             450318221   \n",
       "4  PRJNA837750  SAMN28230094        MIGS.ba  780487290; 518534252   \n",
       "\n",
       "                          Center Name Collection_Date  ...  \\\n",
       "0  UNIVERSITY OF CALIFORNIA SAN DIEGO      2020-08-20  ...   \n",
       "1  UNIVERSITY OF CALIFORNIA SAN DIEGO      2020-08-20  ...   \n",
       "2  UNIVERSITY OF CALIFORNIA SAN DIEGO      2020-08-20  ...   \n",
       "3  UNIVERSITY OF CALIFORNIA SAN DIEGO      2020-08-20  ...   \n",
       "4  UNIVERSITY OF CALIFORNIA SAN DIEGO      2020-08-20  ...   \n",
       "\n",
       "                                  create_date version Sample Name  SRA Study  \\\n",
       "0                        2022-05-17T08:03:00Z       1      CCS169  SRP375596   \n",
       "1  2022-05-17T08:07:00Z; 2022-05-17T08:04:00Z       1      CCS183  SRP375598   \n",
       "2  2022-05-16T18:28:00Z; 2022-05-16T18:23:00Z       1      CCS174  SRP375504   \n",
       "3                        2022-05-16T18:22:00Z       1      CCS178  SRP375502   \n",
       "4  2022-05-16T18:28:00Z; 2022-05-16T18:23:00Z       1      CCS156  SRP375506   \n",
       "\n",
       "   strain    Cheese batch Depth elev week  \n",
       "0  CCS169  cheese_B   NaN   NaN  NaN  NaN  \n",
       "1  CCS183  cheese_B   NaN   NaN  NaN  NaN  \n",
       "2  CCS174  cheese_B   NaN   NaN  NaN  NaN  \n",
       "3  CCS178  cheese_B   NaN   NaN  NaN  NaN  \n",
       "4  CCS156  cheese_B   NaN   NaN  NaN  NaN  \n",
       "\n",
       "[5 rows x 43 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# add prefix cheese_ to the rows of Cheese column in Saak2023_metagenome\n",
    "Saak2023_metagenome['Cheese'] = 'cheese_' + Saak2023_metagenome['Cheese']\n",
    "\n",
    "# Add column Cheese to Saak2023_metadata all values are cheese_B\n",
    "Saak2023_metadata['Cheese'] = 'cheese_B'\n",
    "\n",
    "# concatenate the two dataframes\n",
    "Saak2023_combined = pd.concat([Saak2023_metadata, Saak2023_metagenome], ignore_index=True)\n",
    "\n",
    "display(Saak2023_combined.head())\n",
    "# save the combined dataframe to a file\n",
    "Saak2023_combined.to_csv('data/Saak2023_combined_metadata.tsv', sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>mag_id</th>\n",
       "      <th>sample_description</th>\n",
       "      <th>fermented_food</th>\n",
       "      <th>specific_substrate</th>\n",
       "      <th>substrate_category</th>\n",
       "      <th>general_category</th>\n",
       "      <th>completeness</th>\n",
       "      <th>contamination</th>\n",
       "      <th>contigs</th>\n",
       "      <th>total_length</th>\n",
       "      <th>...</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>order</th>\n",
       "      <th>family</th>\n",
       "      <th>genus</th>\n",
       "      <th>species</th>\n",
       "      <th>study_catalog</th>\n",
       "      <th>source</th>\n",
       "      <th>dataset_name</th>\n",
       "      <th>sample_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AlvarezOrdonezA_xxxx__C02-18__bin.2</td>\n",
       "      <td>meat</td>\n",
       "      <td>sausage</td>\n",
       "      <td>meat</td>\n",
       "      <td>meat</td>\n",
       "      <td>fermented_meat</td>\n",
       "      <td>51.600000</td>\n",
       "      <td>3.100000</td>\n",
       "      <td>2506</td>\n",
       "      <td>20998336</td>\n",
       "      <td>...</td>\n",
       "      <td>Ascomycota</td>\n",
       "      <td>Eurotiomycetes</td>\n",
       "      <td>Eurotiales</td>\n",
       "      <td>Aspergillaceae</td>\n",
       "      <td>Aspergillus</td>\n",
       "      <td>Aspergillus_glaucus</td>\n",
       "      <td>Carlino2024</td>\n",
       "      <td>AlvarezOrdonezA_xxxx__C02-18</td>\n",
       "      <td>AlvarezOrdonezA_xxxx</td>\n",
       "      <td>C02-18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AlvarezOrdonezA_xxxx__C02-19__bin.5</td>\n",
       "      <td>meat</td>\n",
       "      <td>sausage</td>\n",
       "      <td>meat</td>\n",
       "      <td>meat</td>\n",
       "      <td>fermented_meat</td>\n",
       "      <td>63.793103</td>\n",
       "      <td>0.862069</td>\n",
       "      <td>542</td>\n",
       "      <td>1596974</td>\n",
       "      <td>...</td>\n",
       "      <td>Firmicutes</td>\n",
       "      <td>Bacilli</td>\n",
       "      <td>Bacillales</td>\n",
       "      <td>Staphylococcaceae</td>\n",
       "      <td>Staphylococcus</td>\n",
       "      <td>Staphylococcus_equorum</td>\n",
       "      <td>Carlino2024</td>\n",
       "      <td>AlvarezOrdonezA_xxxx__C02-19</td>\n",
       "      <td>AlvarezOrdonezA_xxxx</td>\n",
       "      <td>C02-19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AlvarezOrdonezA_xxxx__C02-19__bin.3</td>\n",
       "      <td>meat</td>\n",
       "      <td>sausage</td>\n",
       "      <td>meat</td>\n",
       "      <td>meat</td>\n",
       "      <td>fermented_meat</td>\n",
       "      <td>74.137931</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>279</td>\n",
       "      <td>5340756</td>\n",
       "      <td>...</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Pseudomonadales</td>\n",
       "      <td>Pseudomonadaceae</td>\n",
       "      <td>Pseudomonas</td>\n",
       "      <td>Pseudomonas_lactis</td>\n",
       "      <td>Carlino2024</td>\n",
       "      <td>AlvarezOrdonezA_xxxx__C02-19</td>\n",
       "      <td>AlvarezOrdonezA_xxxx</td>\n",
       "      <td>C02-19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AlvarezOrdonezA_xxxx__C02-19__bin.1</td>\n",
       "      <td>meat</td>\n",
       "      <td>sausage</td>\n",
       "      <td>meat</td>\n",
       "      <td>meat</td>\n",
       "      <td>fermented_meat</td>\n",
       "      <td>64.912281</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>712</td>\n",
       "      <td>3816144</td>\n",
       "      <td>...</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Enterobacterales</td>\n",
       "      <td>Yersiniaceae</td>\n",
       "      <td>Ewingella</td>\n",
       "      <td>Ewingella_americana</td>\n",
       "      <td>Carlino2024</td>\n",
       "      <td>AlvarezOrdonezA_xxxx__C02-19</td>\n",
       "      <td>AlvarezOrdonezA_xxxx</td>\n",
       "      <td>C02-19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AlvarezOrdonezA_xxxx__C02-19__bin.10</td>\n",
       "      <td>meat</td>\n",
       "      <td>sausage</td>\n",
       "      <td>meat</td>\n",
       "      <td>meat</td>\n",
       "      <td>fermented_meat</td>\n",
       "      <td>63.900000</td>\n",
       "      <td>3.300000</td>\n",
       "      <td>1188</td>\n",
       "      <td>9316757</td>\n",
       "      <td>...</td>\n",
       "      <td>Ascomycota</td>\n",
       "      <td>Saccharomycetes</td>\n",
       "      <td>Saccharomycetales</td>\n",
       "      <td>Debaryomycetaceae</td>\n",
       "      <td>Debaryomyces</td>\n",
       "      <td>Debaryomyces_tyrocola</td>\n",
       "      <td>Carlino2024</td>\n",
       "      <td>AlvarezOrdonezA_xxxx__C02-19</td>\n",
       "      <td>AlvarezOrdonezA_xxxx</td>\n",
       "      <td>C02-19</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 23 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                 mag_id sample_description fermented_food  \\\n",
       "0   AlvarezOrdonezA_xxxx__C02-18__bin.2               meat        sausage   \n",
       "1   AlvarezOrdonezA_xxxx__C02-19__bin.5               meat        sausage   \n",
       "2   AlvarezOrdonezA_xxxx__C02-19__bin.3               meat        sausage   \n",
       "3   AlvarezOrdonezA_xxxx__C02-19__bin.1               meat        sausage   \n",
       "4  AlvarezOrdonezA_xxxx__C02-19__bin.10               meat        sausage   \n",
       "\n",
       "  specific_substrate substrate_category general_category  completeness  \\\n",
       "0               meat               meat   fermented_meat     51.600000   \n",
       "1               meat               meat   fermented_meat     63.793103   \n",
       "2               meat               meat   fermented_meat     74.137931   \n",
       "3               meat               meat   fermented_meat     64.912281   \n",
       "4               meat               meat   fermented_meat     63.900000   \n",
       "\n",
       "   contamination  contigs  total_length  ...          phylum  \\\n",
       "0       3.100000     2506      20998336  ...      Ascomycota   \n",
       "1       0.862069      542       1596974  ...      Firmicutes   \n",
       "2       0.000000      279       5340756  ...  Proteobacteria   \n",
       "3       0.000000      712       3816144  ...  Proteobacteria   \n",
       "4       3.300000     1188       9316757  ...      Ascomycota   \n",
       "\n",
       "                 class              order             family           genus  \\\n",
       "0       Eurotiomycetes         Eurotiales     Aspergillaceae     Aspergillus   \n",
       "1              Bacilli         Bacillales  Staphylococcaceae  Staphylococcus   \n",
       "2  Gammaproteobacteria    Pseudomonadales   Pseudomonadaceae     Pseudomonas   \n",
       "3  Gammaproteobacteria   Enterobacterales       Yersiniaceae       Ewingella   \n",
       "4      Saccharomycetes  Saccharomycetales  Debaryomycetaceae    Debaryomyces   \n",
       "\n",
       "                  species study_catalog                        source  \\\n",
       "0     Aspergillus_glaucus   Carlino2024  AlvarezOrdonezA_xxxx__C02-18   \n",
       "1  Staphylococcus_equorum   Carlino2024  AlvarezOrdonezA_xxxx__C02-19   \n",
       "2      Pseudomonas_lactis   Carlino2024  AlvarezOrdonezA_xxxx__C02-19   \n",
       "3     Ewingella_americana   Carlino2024  AlvarezOrdonezA_xxxx__C02-19   \n",
       "4   Debaryomyces_tyrocola   Carlino2024  AlvarezOrdonezA_xxxx__C02-19   \n",
       "\n",
       "           dataset_name sample_id  \n",
       "0  AlvarezOrdonezA_xxxx    C02-18  \n",
       "1  AlvarezOrdonezA_xxxx    C02-19  \n",
       "2  AlvarezOrdonezA_xxxx    C02-19  \n",
       "3  AlvarezOrdonezA_xxxx    C02-19  \n",
       "4  AlvarezOrdonezA_xxxx    C02-19  \n",
       "\n",
       "[5 rows x 23 columns]"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Split source column into dataset_name and sample_id\n",
    "dataset[['dataset_name', 'sample_id']] = dataset['mag_id'].str.rsplit('__', n=2, expand=True).iloc[:, :2]\n",
    "dataset.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(13858, 23)"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dataset.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "No duplicates found in carlino_metadata\n"
     ]
    }
   ],
   "source": [
    "# Check for duplicates in carlino_metadata using dataset_name and sample_id as keys\n",
    "duplicates = carlino_metadata[carlino_metadata.duplicated(['dataset_name', 'sample_id'], keep=False)]\n",
    "\n",
    "# Display counts of duplicates if any exist\n",
    "if len(duplicates) > 0:\n",
    "    print(f\"Found {len(duplicates)} duplicate entries\")\n",
    "    # Group duplicates and show their counts\n",
    "    duplicate_counts = duplicates.groupby(['dataset_name', 'sample_id']).size().reset_index(name='count')\n",
    "    display(duplicate_counts)\n",
    "    \n",
    "    # Display the actual duplicate rows\n",
    "    print(\"\\nDuplicate rows:\")\n",
    "    display(duplicates)\n",
    "else:\n",
    "    print(\"No duplicates found in carlino_metadata\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of duplicates after removal: 0\n"
     ]
    }
   ],
   "source": [
    "# Drop duplicates keeping the first occurrence\n",
    "carlino_metadata = carlino_metadata.drop_duplicates(['sample_id'], keep='first')\n",
    "\n",
    "# Verify the duplicate is removed\n",
    "print(\"Number of duplicates after removal:\",\n",
    "    carlino_metadata.duplicated(['dataset_name', 'sample_id']).sum())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(3366, 33)"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "dataset_name                 85\n",
       "sample_id                  3366\n",
       "macrocategory                 1\n",
       "category                     15\n",
       "type                        118\n",
       "subtype                     337\n",
       "commercial_name              13\n",
       "fermented/non-fermented       2\n",
       "country                      52\n",
       "sample_accession           3303\n",
       "run_accession              3196\n",
       "experiment_accession       1367\n",
       "study_accession              85\n",
       "project_accession            85\n",
       "database_origin               4\n",
       "library_layout                3\n",
       "sequencing_platform          16\n",
       "DNA_extraction_kit           35\n",
       "collection_date             201\n",
       "n_of_bases                 3366\n",
       "n_of_reads                 3364\n",
       "min_read_len                 33\n",
       "median_read_len              62\n",
       "mean_read_len              3027\n",
       "max_read_len                 26\n",
       "n_contigs                  3065\n",
       "n_MAGs_MQ_prok               17\n",
       "n_MAGs_HQ_prok               18\n",
       "n_MAGs_MQ_euk                 7\n",
       "n_MAGs_HQ_euk                 6\n",
       "filtered                      2\n",
       "curator                       4\n",
       "Reference                    34\n",
       "dtype: int64"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# check the number of unique values in each column of carlino_metadata\n",
    "display(carlino_metadata.shape, carlino_metadata.nunique())\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Shape after merge:\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "(13858, 38)"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "(13858, 23)"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Merge dataset with selected columns from carlino_metadata\n",
    "merged_df = pd.merge(dataset, \n",
    "                    carlino_metadata.iloc[:, :17], \n",
    "                    how='left',\n",
    "                    on=['dataset_name', 'sample_id'])\n",
    "\n",
    "# Verify the merge\n",
    "print(\"\\nShape after merge:\")\n",
    "display(merged_df.shape, dataset.shape)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>count</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>study_catalog</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Du2023</th>\n",
       "      <td>553</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Caffrey2024</th>\n",
       "      <td>350</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Saak2023_CheeseMAGs</th>\n",
       "      <td>52</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Rappaport2024_sourdough</th>\n",
       "      <td>29</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Carlino2024</th>\n",
       "      <td>23</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                         count\n",
       "study_catalog                 \n",
       "Du2023                     553\n",
       "Caffrey2024                350\n",
       "Saak2023_CheeseMAGs         52\n",
       "Rappaport2024_sourdough     29\n",
       "Carlino2024                 23"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Total number of unique study_catalog values for samples without sequencing platform: 5\n",
      "\n",
      "Total samples without sequencing platform: 1007\n"
     ]
    }
   ],
   "source": [
    "#print the unique values of columns study_catalog for data that was not found in carlino\n",
    "# Get value counts of study_catalog for entries with missing sequencing_platform\n",
    "study_counts = merged_df[merged_df['sequencing_platform'].isna()]['study_catalog'].value_counts().to_frame()\n",
    "display(study_counts)\n",
    "print(f\"\\nTotal number of unique study_catalog values for samples without sequencing platform: {len(study_counts)}\")\n",
    "\n",
    "# Print total number of rows with missing sequencing_platform\n",
    "print(f\"\\nTotal samples without sequencing platform: {merged_df['sequencing_platform'].isna().sum()}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
